{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "29f9c521-f794-4da6-a44f-ed03ee62be2f",
   "metadata": {},
   "source": [
    "# Trying_21_stuff_in_SICILIAN_via_MYBinder TO BECOME Demo_SICILIAN_via_MyBinder_with_human_Chr_21\n",
    "\n",
    "If you didn't start your Jupyter session [here](https://github.com/fomightez/SICILIAN-binder), go there and use '`launch binder`' badge to get started.\n",
    "\n",
    "Here, in the demonstration Human chromosome 21 will serve as the input 'genome'. Chr 21 is the smallest human chromosome according to the telomere-to-telomere data, see [here](https://ncbi.nlm.nih.gov/datasets/genome/GCF_009914755.1 ). \n",
    "\n",
    "We are only using a fraction of the human genome and small human scRNA-seq (Smart-seq2) dataset here for this demonstration in the interest of time, to help keep the demonstration interactive, and **because the computational resources offered by MyBinder are limited**. (Specifics on the human Smart-seq2 data: [GEO Sample GSM4678248](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM4678248) -- TF-driven_differentiated_hiPSC 01_A01, which is one of 95 fastq files provided in [GEO series GSE154759](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE154759) corresponding to each individual sorted cell; with the reads accessible in [Run SRR12269911](https://trace.ncbi.nlm.nih.gov/Traces/?view=run_browser&acc=SRR12269911&display=metadata) of [entry accesion SRX8774970](https://www.ncbi.nlm.nih.gov/sra?term=SRX8774970) at the Short Read Archive;ref. [Ng et al., 2021](https://pubmed.ncbi.nlm.nih.gov/33257861/).) While limited, having an environment offering a working pipeline can be helpful to have handy when learning about the pipeline and trying to work out how to use the SICILIAN pipeline in places with more power.\n",
    "\n",
    "We'll need to index human chromsome 21 with STAR aligner so SICILIAN can use that. Additional input files and the reads from one run of an scRNA-seq analysis are already present in the session provided upon launch.\n",
    "\n",
    "Overview of this notebook:\n",
    "\n",
    "1. Install STAR aligner and verify it works to make a mock read sequence to a mock, tiny genome.\n",
    "1. Index human chromosome 21 with STAR aligner.\n",
    "1. Make supporting annotator files.\n",
    "1. Configure the main script of SICILIAN, `SICILIAN.py` to use the generated index, generated supportingfiles, and the provided set of reads from a Smart-seq2 scRNA-seq analysis.\n",
    "1. Edit `SICILIAN.py` to not use SLURM.\n",
    "1. For now use the `light_utils.py` provided by the authors (this is an atypical step that you won't hopefully need soon)\n",
    "1. Run the SICILIAN pipeline"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc23e670-2be6-4a36-90c4-8bdfa4a0b68d",
   "metadata": {},
   "source": [
    "### Preparation STEP #1: Install STAR aligner, which is a dependency of SICILIAN"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7187ffa3-3481-4ae4-8ac1-1506e280f3e8",
   "metadata": {
    "tags": []
   },
   "source": [
    "Install the STAR Aligner to prepare for using SICILIAN.\n",
    "\n",
    "##### Technical note on environment set up for SICILIAN use  \n",
    "STAR Aligner could have been specified to be installed already in MyBinder-provided sessions that I assume you are using while reading this. I chose not to do that though as setting up the environment for SICILIAN is quite a challenge as it involves Python, R, and other softwware, packages and libraries in support and I wanted to have at least one representative dependency that still needed to be installed as a representative of that. The `environment.yml` [here](https://github.com/fomightez/SICILIAN-binder/blob/master/binder/environment.yml) should give you a sense of this complexity.       \n",
    "After installing STAR aligner, we'll establish it works by very quickly mapping a single mock sequence read to a tiny mock genome.\n",
    "That will set us up for further preparation steps and for ultimately running the SICILIAN pipeline."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c0e7d808-1a4c-441a-b164-6fa16fc2c7b9",
   "metadata": {},
   "source": [
    "##### Installing STAR aligner here\n",
    "\n",
    "STAR aligner will be installed here with the following command run in a cell:\n",
    "\n",
    "```shell\n",
    "%conda install -y bioconda::star\n",
    "```\n",
    "\n",
    "If you were doing this in a terminal, you'd leave off the magic `%` symbol. That magic `%` symbol combined with the typical install command, makes a special magic install command in the context of Jupyter that makes sure the installation occurs where the kernel backing this active notebook is running, making managing environments more convenient.\n",
    "\n",
    "To actually handle the installation of STAR aligner and verify it is working by mapping a read to a mock genome, we can run the Jupyter notebook file `STAR_aligner_run_on_MyBinder.ipynb` with the command `%run STAR_aligner_run_on_MyBinder.ipynb` shown in the cell below. I have added printing the steps to stderr (the pink highlighted lines) to better enable understanding the context of the steps without needing to actually open the notebook `STAR_aligner_run_on_MyBinder.ipynb`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "928fb930-4a8f-430f-9f3b-edbea2ad63f3",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Install STAR aligner to prepare\n",
      "Install is done via the command `%conda install -y bioconda::star`\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Retrieving notices: ...working... done\n",
      "Channels:\n",
      " - conda-forge\n",
      " - bioconda\n",
      "Platform: linux-64\n",
      "Collecting package metadata (repodata.json): done\n",
      "Solving environment: done\n",
      "\n",
      "\n",
      "==> WARNING: A newer version of conda exists. <==\n",
      "    current version: 24.5.0\n",
      "    latest version: 24.7.1\n",
      "\n",
      "Please update conda by running\n",
      "\n",
      "    $ conda update -n base -c conda-forge conda\n",
      "\n",
      "\n",
      "\n",
      "## Package Plan ##\n",
      "\n",
      "  environment location: /srv/conda/envs/notebook\n",
      "\n",
      "  added / updated specs:\n",
      "    - bioconda::star\n",
      "\n",
      "\n",
      "The following packages will be downloaded:\n",
      "\n",
      "    package                    |            build\n",
      "    ---------------------------|-----------------\n",
      "    htslib-1.21                |       h5efdd21_0         3.0 MB  bioconda\n",
      "    star-2.7.11b               |       h43eeafb_2         8.1 MB  bioconda\n",
      "    ------------------------------------------------------------\n",
      "                                           Total:        11.1 MB\n",
      "\n",
      "The following NEW packages will be INSTALLED:\n",
      "\n",
      "  htslib             bioconda/linux-64::htslib-1.21-h5efdd21_0 \n",
      "  star               bioconda/linux-64::star-2.7.11b-h43eeafb_2 \n",
      "\n",
      "\n",
      "\n",
      "Downloading and Extracting Packages:\n",
      "star-2.7.11b         | 8.1 MB    |                                       |   0% \n",
      "htslib-1.21          | 3.0 MB    |                                       |   0% \u001b[A\n",
      "star-2.7.11b         | 8.1 MB    |                                       |   0% \u001b[A\n",
      "star-2.7.11b         | 8.1 MB    | ##5                                   |   7% \u001b[A\n",
      "star-2.7.11b         | 8.1 MB    | ####3                                 |  12% \u001b[A\n",
      "star-2.7.11b         | 8.1 MB    | ######4                               |  17% \u001b[A\n",
      "star-2.7.11b         | 8.1 MB    | ##########7                           |  29% \u001b[A\n",
      "star-2.7.11b         | 8.1 MB    | ############7                         |  34% \u001b[A\n",
      "star-2.7.11b         | 8.1 MB    | ##################################5   |  93% \u001b[A\n",
      "                                                                                \u001b[A\n",
      "                                                                                \u001b[A\n",
      "Preparing transaction: done\n",
      "Verifying transaction: done\n",
      "Executing transaction: done\n",
      "\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Run a test alignment to demonstrate it works\n",
      "1. Create a small FASTA file with a short sequence:\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">test_seq\n",
      "ACGTGGACGTACCCGTACGTACCCAAGTACGAGTACGTACGGGTACGTTCCACAAGTACGT\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2. Run `mkdir test_index` to make the directory to store the index."
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "drwxr-xr-x 2 jovyan jovyan 4.0K Sep 26 17:32 \u001b[0m\u001b[01;34mtest_index\u001b[0m/\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "3. Generate a genome index using this small FASTA file:\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\t/srv/conda/envs/notebook/bin/STAR-avx2 --runMode genomeGenerate --genomeFastaFiles test.fa --genomeDir test_index --genomeSAindexNbases 1\n",
      "\tSTAR version: 2.7.11b   compiled: 2024-07-03T14:39:20+0000 :/opt/conda/conda-bld/star_1720017372352/work/source\n",
      "Sep 26 17:32:09 ..... started STAR run\n",
      "Sep 26 17:32:09 ... starting to generate Genome files\n",
      "Sep 26 17:32:09 ... starting to sort Suffix Array. This may take a long time...\n",
      "Sep 26 17:32:09 ... sorting Suffix Array chunks and saving them to disk...\n",
      "Sep 26 17:32:09 ... loading chunks from disk, packing SA...\n",
      "Sep 26 17:32:09 ... finished generating suffix array\n",
      "Sep 26 17:32:09 ... generating Suffix Array index\n",
      "Sep 26 17:32:09 ... completed Suffix Array index\n",
      "Sep 26 17:32:09 ... writing Genome to disk ...\n",
      "Sep 26 17:32:09 ... writing Suffix Array to disk ...\n",
      "Sep 26 17:32:09 ... writing SAindex to disk\n",
      "Sep 26 17:32:09 ..... finished successfully\n",
      "\t/srv/conda/envs/notebook/bin/STAR-avx2 --runMode alignReads --genomeDir test_index --readFilesIn test.fq --outFileNamePrefix test_output\n",
      "\tSTAR version: 2.7.11b   compiled: 2024-07-03T14:39:20+0000 :/opt/conda/conda-bld/star_1720017372352/work/source\n",
      "Sep 26 17:32:09 ..... started STAR run\n",
      "Sep 26 17:32:09 ..... loading genome\n",
      "Sep 26 17:32:09 ..... started mapping\n",
      "Sep 26 17:32:09 ..... finished mapping\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "4. Create a small FASTQ file with a few reads from the test sequence:\n",
      "5. Run the alignment using the small test files:\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sep 26 17:32:09 ..... finished successfully\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "SHOW THE FILES GENERATED BY ALL THAT AND THE ALIGNING STEP.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 468K\n",
      "drwxr-x--- 1 jovyan jovyan 4.0K Sep 26 17:32 \u001b[0m\u001b[01;34m.\u001b[0m/\n",
      "drwxr-xr-x 1 root   root   4.0K Sep 19 20:15 \u001b[01;34m..\u001b[0m/\n",
      "-rw-r--r-- 1 jovyan jovyan  220 Jan  6  2022 .bash_logout\n",
      "-rw-r--r-- 1 jovyan jovyan 3.7K Jan  6  2022 .bashrc\n",
      "drwxr-xr-x 3 jovyan jovyan 4.0K Sep 19 20:09 \u001b[01;34mbenchmarking\u001b[0m/\n",
      "drwxr-xr-x 1 jovyan jovyan 4.0K Sep 19 20:09 \u001b[01;34mbinder\u001b[0m/\n",
      "drwxr-xr-x 1 jovyan jovyan 4.0K Sep 19 20:13 \u001b[01;34m.cache\u001b[0m/\n",
      "drwxrwsr-x 1 jovyan jovyan 4.0K Sep 19 20:13 \u001b[01;34m.conda\u001b[0m/\n",
      "drwxr-xr-x 2 jovyan jovyan 4.0K Sep 19 20:09 \u001b[01;34mdemo_input\u001b[0m/\n",
      "-rw-r--r-- 1 jovyan jovyan  54K Sep 26 17:20 Demo_SICILIAN_via_MyBinder_with_human_Chr_21.ipynb\n",
      "drwxr-xr-x 8 jovyan jovyan 4.0K Sep 19 20:09 \u001b[01;34m.git\u001b[0m/\n",
      "-rw-r--r-- 1 jovyan jovyan   22 Sep 19 20:09 .gitignore\n",
      "drwxr-xr-x 2 jovyan jovyan 4.0K Sep 26 17:15 \u001b[01;34m.ipynb_checkpoints\u001b[0m/\n",
      "drwxr-xr-x 1 jovyan jovyan 4.0K Sep 26 16:08 \u001b[01;34m.ipython\u001b[0m/\n",
      "drwxr-xr-x 3 jovyan jovyan 4.0K Sep 26 16:08 \u001b[01;34m.jupyter\u001b[0m/\n",
      "-rw-r--r-- 1 jovyan jovyan 8.1K Sep 26 17:30 .jupyter-server-log.txt\n",
      "-rw-r--r-- 1 jovyan jovyan  18K Sep 19 20:09 LICENSE\n",
      "drwxr-xr-x 1 jovyan jovyan 4.0K Sep 19 20:16 \u001b[01;34m.local\u001b[0m/\n",
      "drwxr-xr-x 3 jovyan jovyan 4.0K Sep 26 16:08 \u001b[01;34m.npm\u001b[0m/\n",
      "-rw-r--r-- 1 jovyan jovyan  807 Jan  6  2022 .profile\n",
      "-rw-r--r-- 1 jovyan jovyan  26K Sep 19 20:09 README.md\n",
      "-rw-r--r-- 1 jovyan jovyan   89 Sep 19 20:09 requirements.txt\n",
      "-rw-r--r-- 1 jovyan jovyan  38K Sep 26 16:11 Run_SICILIAN_demo_on_MyBinder.ipynb\n",
      "-rw-r--r-- 1 jovyan jovyan 1.5K Sep 19 20:09 RUN_SICILIAN_post_process.sh\n",
      "drwxr-xr-x 3 jovyan jovyan 4.0K Sep 19 20:09 \u001b[01;34mscripts\u001b[0m/\n",
      "-rw-r--r-- 1 jovyan jovyan 141K Sep 19 20:09 SICILIAN.png\n",
      "-rw-r--r-- 1 jovyan jovyan 6.7K Sep 19 20:09 SICILIAN_post_process.py\n",
      "-rw-r--r-- 1 jovyan jovyan 9.8K Sep 19 20:09 SICILIAN.py\n",
      "-rw-r--r-- 1 jovyan jovyan  20K Sep 19 20:09 STAR_aligner_run_on_MyBinder.ipynb\n",
      "-rw-r--r-- 1 jovyan jovyan   72 Sep 26 17:32 test.fa\n",
      "-rw-r--r-- 1 jovyan jovyan   49 Sep 26 17:32 test.fq\n",
      "drwxr-xr-x 2 jovyan jovyan 4.0K Sep 26 17:32 \u001b[01;34mtest_index\u001b[0m/\n",
      "-rw-r--r-- 1 jovyan jovyan  481 Sep 26 17:32 test_outputAligned.out.sam\n",
      "-rw-r--r-- 1 jovyan jovyan 2.0K Sep 26 17:32 test_outputLog.final.out\n",
      "-rw-r--r-- 1 jovyan jovyan 3.7K Sep 26 17:32 test_outputLog.out\n",
      "-rw-r--r-- 1 jovyan jovyan  246 Sep 26 17:32 test_outputLog.progress.out\n",
      "-rw-r--r-- 1 jovyan jovyan    0 Sep 26 17:32 test_outputSJ.out.tab\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Examine results to see it worked\n",
      "Results in 'test_outputLog.final.out'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                 Started job on |\tSep 26 17:32:09\n",
      "                             Started mapping on |\tSep 26 17:32:09\n",
      "                                    Finished on |\tSep 26 17:32:09\n",
      "       Mapping speed, Million of reads per hour |\tinf\n",
      "\n",
      "                          Number of input reads |\t1\n",
      "                      Average input read length |\t19\n",
      "                                    UNIQUE READS:\n",
      "                   Uniquely mapped reads number |\t1\n",
      "                        Uniquely mapped reads % |\t100.00%\n",
      "                          Average mapped length |\t19.00\n",
      "                       Number of splices: Total |\t0\n",
      "            Number of splices: Annotated (sjdb) |\t0\n",
      "                       Number of splices: GT/AG |\t0\n",
      "                       Number of splices: GC/AG |\t0\n",
      "                       Number of splices: AT/AC |\t0\n",
      "               Number of splices: Non-canonical |\t0\n",
      "                      Mismatch rate per base, % |\t0.00%\n",
      "                         Deletion rate per base |\t0.00%\n",
      "                        Deletion average length |\t0.00\n",
      "                        Insertion rate per base |\t0.00%\n",
      "                       Insertion average length |\t0.00\n",
      "                             MULTI-MAPPING READS:\n",
      "        Number of reads mapped to multiple loci |\t0\n",
      "             % of reads mapped to multiple loci |\t0.00%\n",
      "        Number of reads mapped to too many loci |\t0\n",
      "             % of reads mapped to too many loci |\t0.00%\n",
      "                                  UNMAPPED READS:\n",
      "  Number of reads unmapped: too many mismatches |\t0\n",
      "       % of reads unmapped: too many mismatches |\t0.00%\n",
      "            Number of reads unmapped: too short |\t0\n",
      "                 % of reads unmapped: too short |\t0.00%\n",
      "                Number of reads unmapped: other |\t0\n",
      "                     % of reads unmapped: other |\t0.00%\n",
      "                                  CHIMERIC READS:\n",
      "                       Number of chimeric reads |\t0\n",
      "                            % of chimeric reads |\t0.00%\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Results in 'test_outputAligned.out.sam'"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "@HD\tVN:1.4\n",
      "@SQ\tSN:test_seq\tLN:61\n",
      "@PG\tID:STAR\tPN:STAR\tVN:2.7.11b\tCL:/srv/conda/envs/notebook/bin/STAR-avx2   --runMode alignReads      --genomeDir test_index   --readFilesIn test.fq      --outFileNamePrefix test_output\n",
      "@CO\tuser command line: /srv/conda/envs/notebook/bin/STAR-avx2 --runMode alignReads --genomeDir test_index --readFilesIn test.fq --outFileNamePrefix test_output\n",
      "read1\t0\ttest_seq\t19\t255\t19M\t*\t0\t0\tGTACCCAAGTACGAGTACG\tAAAAAAAAAAAAAAAAAAA\tNH:i:1\tHI:i:1\tAS:i:18\tnM:i:0\n"
     ]
    }
   ],
   "source": [
    "%run STAR_aligner_run_on_MyBinder.ipynb"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "22ced2d9-b0e9-4cb6-8b67-2f030f0086c6",
   "metadata": {},
   "source": [
    "### Preparation STEP #2: Index human chromosome 21\n",
    "\n",
    "Now that we know that the STAR aligner software is installed and works, we'll index human chromsome 21 with STAR aligner so that the SICILAN pipeline can use that generated indexing in the process of mapping reads to human chromsome 21 later. Recall, human chromosome 21 is serving as our 'genome' in this limited computational environment. The small size also helps have the demonstration run in a reasonable amount of time for it to be interactive, which wouldn't be expected to the case for a true run of the SICILIAN pipeline, or at least some steps.    \n",
    "The `star_ref_path` will later be set in the configuration settings to point at the generated indexing results.\n",
    "\n",
    "We can run the following command to generate that index here:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2eabed5d-adbf-411b-b869-b7491d128da4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\t/srv/conda/envs/notebook/bin/STAR-avx2 --runMode genomeGenerate --genomeFastaFiles ./demo_input/human_chromosome_21_hg38.fasta --genomeDir human21_index --genomeSAindexNbases 11\n",
      "\tSTAR version: 2.7.11b   compiled: 2024-07-03T14:39:20+0000 :/opt/conda/conda-bld/star_1720017372352/work/source\n",
      "Sep 26 17:32:11 ..... started STAR run\n",
      "Sep 26 17:32:11 ... starting to generate Genome files\n",
      "Sep 26 17:32:13 ... starting to sort Suffix Array. This may take a long time...\n",
      "Sep 26 17:32:13 ... sorting Suffix Array chunks and saving them to disk...\n",
      "Sep 26 17:33:37 ... loading chunks from disk, packing SA...\n",
      "Sep 26 17:33:39 ... finished generating suffix array\n",
      "Sep 26 17:33:39 ... generating Suffix Array index\n",
      "Sep 26 17:33:43 ... completed Suffix Array index\n",
      "Sep 26 17:33:43 ... writing Genome to disk ...\n",
      "Sep 26 17:33:43 ... writing Suffix Array to disk ...\n",
      "Sep 26 17:33:44 ... writing SAindex to disk\n",
      "Sep 26 17:33:44 ..... finished successfully\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "STAR --runMode genomeGenerate \\\n",
    "     --genomeFastaFiles \"./demo_input/human_chromosome_21_hg38.fasta\" \\\n",
    "     --genomeDir human21_index \\\n",
    "     --genomeSAindexNbases 11"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42089ef0-fc31-4d71-87ce-b4ccb319583d",
   "metadata": {},
   "source": [
    "`genomeSAindexNbases` was set here based on STAR aligner recommending that for this input chromosome. You'll need to adjust that value based on your genome, see more details including a general forumula [here](https://github.com/alexdobin/STAR/issues/972). "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fac5bcee-fbe3-4596-83a0-cead799ba1b2",
   "metadata": {},
   "source": [
    "### Preparation STEP #3: Use SICILIAN's `create_annotator.py` script to prepare the supporting annotator files\n",
    "\n",
    "Although the SICILIAN algorithm is agnostic to the annotation, in order to indicate how the junctions detected relate the to expected junctions, the SICILIAN pipeline uses supporting files with the details of genes and exons in relation to expected splicing.  \n",
    "SICILIAN's create_annotator.py script uses Python's pickling to store annotations in pickle files. (For context, pickling is a convenient way to save Python objects in a reusable format.)\n",
    "\n",
    "Normally, as part of preparing to run SICILIAN's main script, you'd probably use SICILIAN's `create_annotator.py` script to generate these supporting annotator files. Depending on what you are using though you may not have to generate them yourself. Fortunately, that the SICILIAN authors provide pre-computed, ready-to-use supporting index and annotator files for SICILIAN for a handful of species, see the links listed in the [SICILIAN GitHub repo README under 'Download ready-to-use index and annotator files'](https://github.com/salzman-lab/SICILIAN?tab=readme-ov-file#download-ready-to-use-index-and-annotator-files).  \n",
    "\n",
    "\n",
    "The supporting annotator files need to be made only once by the `create_annotator.py` script in the SICILIAN scripts directory.  \n",
    "\n",
    "The input for  SICILIAN's `create_annotator.py` script is the annotation file for your genome.  \n",
    "The GTF-fromatted annotation file for Chromosome 21 was made by filtering the file `grch38_known_genes.gtf` for only the lines that pertain to Chromosome 21 using `awk`. (`grch38_known_genes.gtf` can be found in `SICILIAN_human_hg38_Refs/gtf_file` among the human hg38 files provided by the SICILIAN authors [here](https://github.com/salzman-lab/SICILIAN?tab=readme-ov-file#download-ready-to-use-index-and-annotator-files).)  The resuling file is provided in the `demo_input` directory here to be used as input..\n",
    "\n",
    "For this demo, we'll run the following command to use SICILIAN's `create_annotator.py` script with human Chromosome 21. \n",
    "\n",
    "With the GTF in hand, now the create_annotator.py script can be run to make the annotator pickle files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "06779b7e-de20-4f3c-9c9c-e38f706f9fd7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dumped exon boundary annotator to /home/jovyan/annotated_21_exon_bounds.pkl\n",
      "dumped splice junctions annotator to /home/jovyan/annotated_21_splices.pkl\n",
      "chr21\n",
      "dumped annotator to /home/jovyan/annotated_21.pkl\n"
     ]
    }
   ],
   "source": [
    "%run scripts/create_annotator.py -g demo_input/grch38_chr21_known_genes.gtf  -a annotated_21"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "89c3c391-dcdf-4d9d-9540-7031acb45253",
   "metadata": {},
   "source": [
    "Note, that in case of issues, as backup the supporting annotator files that should be obtained by running `create_annotator.py` with Chromosome 21 have been provided and can be found in the `demo_input` directory."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69f71d62-b03b-4724-bfc8-a49b372e83a8",
   "metadata": {},
   "source": [
    "### Preparation STEP #4: configure the settings in SICILIAN to reflect all this set up.\n",
    "\n",
    "The sections in `SICILIAN.py` where it says, 'Input arguments that should be set by the user' and 'Toggles for deciding which steps in SICILIAN should be run', lines 145 through 175, need to be edited to match the test data and settings desired. For example, I had thought I didn;t don't need to run STAR aligner again because ran it to set this up and so I planned to skip over by setting `run_map` to `False`. However, I found I needed it to make bam files in location that `SICILIAN.py` expects them and so I ended up not disabling it as I had intended.\n",
    "\n",
    "The string `c` will have the settings I want, and I made a small function to parse it and set them in the current `SICILIAN.py`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "731ded63-ec26-4267-82f1-3f0fcd6b5f9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "c='''###########################################################################################\n",
    "################## Input arguments that should be set by the user  ########################\n",
    "###########################################################################################\n",
    "  data_path = \"./\"\n",
    "  out_dir = \"sicilian_output\"\n",
    "  run_name = \"human_21_hiPSC_data\"\n",
    "  r_ends = [\".fq\",\".fastq\"] #seems you need at least two or get `IndexError: list index out of range` for `command += \"--readFilesIn {}{}{} \".format(data_path, name, r_ends[i])`, but repeating bypasses issue\n",
    "  names = [\"./demo_input/chr21_mapped_SRR12269911\"]\n",
    "  star_path = \"/srv/conda/envs/notebook/bin/STAR\"\n",
    "  star_ref_path = \"./human21_index\"\n",
    "  gtf_file = \"./demo_input/grch38_chr21_known_genes.gtf\"\n",
    "  annotator_file = \"./annotated_21.pkl\"\n",
    "  exon_pickle_file = \"./annotated_21_exon_bounds.pkl\"\n",
    "  splice_pickle_file = \"./annotated_21_splices.pkl\"\n",
    "  domain_file = \"\"\n",
    "  single = True\n",
    "  tenX = False\n",
    "  stranded_library = False\n",
    "  bc_pattern = \"C\"*16 + \"N\"*12\n",
    "#########################################################################################\n",
    "#########################################################################################\n",
    "#########################################################################################\n",
    "\n",
    "\n",
    "## Toggles for deciding which steps in SICILIAN should be run #####\n",
    "  run_whitelist = False\n",
    "  run_extract = False\n",
    "  run_map = True\n",
    "  run_class = True\n",
    "  run_GLM = True\n",
    "###################################################################\n",
    "'''\n",
    "\n",
    "\n",
    "###--------------------------------------------------------------------------------------------------------###\n",
    "#### BELOW IS CODE TO MAKE THE SWAP. DON'T TOUCH BELOW HERE. ONLY MAKE EDITS TO ABOVE.########################\n",
    "import re\n",
    "# Define the pattern to match lines between markers\n",
    "start_point_string = \"Input arguments that should be set by the user\"\n",
    "end_point_string = \"run_GLM = True\"\n",
    "# Specify the file path\n",
    "script_file = \"./SICILIAN.py\"\n",
    "# Read the script file\n",
    "with open(script_file, \"r\") as file:\n",
    "    script_text = file.read()\n",
    "def swap_code_lines_in_SICILIAN_with_string_c(c):\n",
    "    pattern = r\"{}(.*?){}\".format(re.escape(start_point_string), re.escape(end_point_string))\n",
    "    # this next `match` is to set up to remove the evertyhing in front of and including `start_point_string` and `end_point_string` and everything after it from string `c`, so no duplicated borders points\n",
    "    match = re.search(fr\"{re.escape(start_point_string)}(.*?){re.escape(end_point_string)}\", c, re.DOTALL)\n",
    "    if match:\n",
    "        c_from_start_to_end = match.group(1)\n",
    "    else:\n",
    "        c_from_start_to_end = \"\"\n",
    "    replace = f\"{start_point_string}{c_from_start_to_end}{end_point_string}\"\n",
    "    # Replace the matched lines with the new code\n",
    "    new_script_text = re.sub(pattern, replace, script_text, flags=re.DOTALL)\n",
    "    # Write the modified script back to the file\n",
    "    with open(script_file, \"w\") as file:\n",
    "        file.write(new_script_text)\n",
    "swap_code_lines_in_SICILIAN_with_string_c(c)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "345d7bbb-63ef-4c22-8ac9-50b2b0ca1ba6",
   "metadata": {},
   "source": [
    "### Preparation STEP #5: Adjust to run in MyBinder session without Slurm.\n",
    "\n",
    "`SICILIAN.py` comes set up to use `sbatch` to run the jobs. `sbatch` submits a batch script to Slurm, which we don't have in this case.  \n",
    "Luckily, [here](https://github.com/salzman-lab/SICILIAN/issues/8#issuecomment-1059198873) and [here](https://github.com/salzman-lab/SICILIAN/issues/8#issuecomment-1709657122) already worked out how to convert it to use different way to run the associated jobs by replacing `sbatch`.  \n",
    "Running this next cell will do that."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "faea8a3d-61a8-41c0-a77c-2b02695b4ced",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Specify the file path\n",
    "script_file = \"./SICILIAN.py\"\n",
    "# Read the script file\n",
    "with open(script_file, \"r\") as file:\n",
    "    script_text = file.read()\n",
    "original_code_to_replace = '\"sbatch {}\"'\n",
    "replacement = '\"bash {}\"'\n",
    "new_script_text = script_text.replace(original_code_to_replace, replacement)\n",
    "# Write the modified script back to the file\n",
    "with open(script_file, \"w\") as file:\n",
    "    file.write(new_script_text)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c085516-4705-4aac-b9bd-cfedd3f07cbe",
   "metadata": {},
   "source": [
    "### Preparation STEP #6: Adjust the content to use the `light_utils.py` provided by the SICILIAN authors \n",
    "\n",
    "This is not a typical step in a SICILIAN workflow. And this special, added step is only meant to be temporary solution given the state of the code available here at this moment.\n",
    "\n",
    "In this repo, I have been editing `light_utils.py` to bring it up to date with more current Pandas & Numpy so that notificaitons about `SettingWithCopyWarning` and other warnings present in the cours of the 'modify' step won't show up. (I was also hoping to get it to work with data that doesn't include any junctions to aid in testing environments, but that may be put on the backburner.) Because it isn't quite perfected, I suggest at this time to use the [the original `light_utils.py`](https://github.com/salzman-lab/SICILIAN/blob/master/scripts/light_utils.py).\n",
    "\n",
    "Running this next cell will swap out my current draft for [the original `light_utils.py`](https://github.com/salzman-lab/SICILIAN/blob/master/scripts/light_utils.py)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f4287b23-a2c1-4cc5-bfce-dbdc6577fdc6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100 21716  100 21716    0     0  88470      0 --:--:-- --:--:-- --:--:-- 88636\n"
     ]
    }
   ],
   "source": [
    "!curl -OL https://raw.githubusercontent.com/salzman-lab/SICILIAN/refs/heads/master/scripts/light_utils.py\n",
    "!mv light_utils.py scripts/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae7adba2-3983-4999-9c5f-143a296b48e5",
   "metadata": {},
   "source": [
    "-----\n",
    "\n",
    "## Run SICILIAN\n",
    "\n",
    "Run the main SICILIAN script by running the following cell. (If you were on the command line the following command would be more something along the lines of `python SICILIAN.py`. Here I'm using Jupyter's convenient magic that is more feature-rich.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "9745a6ec-07fd-4593-9691-0dff2cc4011d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Thu Sep 26 05:34:01 PM UTC 2024\n",
      "2.7.11b\n",
      "\t/srv/conda/envs/notebook/bin/STAR-avx2 --runThreadN 4 --genomeDir ./human21_index --readFilesIn ././demo_input/chr21_mapped_SRR12269911.fastq --twopassMode Basic --alignIntronMax 1000000 --outFileNamePrefix sicilian_output/human_21_hiPSC_data/./demo_input/chr21_mapped_SRR12269911/2 --outSAMtype BAM Unsorted --outSAMattributes All --chimOutType WithinBAM SoftClip Junctions --chimJunctionOverhangMin 10 --chimSegmentReadGapMax 0 --chimOutJunctionFormat 1 --chimSegmentMin 12 --chimScoreJunctionNonGTAG -4 --chimNonchimScoreDropMin 10 --quantMode GeneCounts --sjdbGTFfile ./demo_input/grch38_chr21_known_genes.gtf --outReadsUnmapped Fastx\n",
      "\tSTAR version: 2.7.11b   compiled: 2024-07-03T14:39:20+0000 :/opt/conda/conda-bld/star_1720017372352/work/source\n",
      "Sep 26 17:34:01 ..... started STAR run\n",
      "Sep 26 17:34:01 ..... loading genome\n",
      "Sep 26 17:34:01 ..... processing annotations GTF\n",
      "Sep 26 17:34:02 ..... inserting junctions into the genome indices\n",
      "Sep 26 17:34:13 ..... started 1st pass mapping\n",
      "Sep 26 17:34:30 ..... finished 1st pass mapping\n",
      "Sep 26 17:34:30 ..... inserting junctions into the genome indices\n",
      "Sep 26 17:34:34 ..... started mapping\n",
      "Sep 26 17:34:59 ..... finished mapping\n",
      "Sep 26 17:34:59 ..... finished successfully\n",
      "Thu Sep 26 05:35:00 PM UTC 2024 (run_map.sh)\n",
      "Thu Sep 26 05:35:00 PM UTC 2024\n",
      "bam_files ['sicilian_output/human_21_hiPSC_data/./demo_input/chr21_mapped_SRR12269911/2Aligned.out.bam']\n",
      "bam_files ['sicilian_output/human_21_hiPSC_data/./demo_input/chr21_mapped_SRR12269911/2Aligned.out.bam']\n",
      "started modify 5.043425798416138\n",
      "/home/jovyan/scripts/light_utils.py:33: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  CI_new[\"geneR1A\"] = CI_new[\"geneR1A\"].fillna(\"\")\n",
      "/home/jovyan/scripts/light_utils.py:34: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  CI_new[\"geneR1B\"] = CI_new[\"geneR1B\"].fillna(\"\")\n",
      "/home/jovyan/scripts/light_utils.py:38: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  CI_new[\"gene_strandR1A\"] = CI_new[\"refName_ABR1\"].str.split(\"|\").str[0].str.split(\":\").str[-1]\n",
      "/home/jovyan/scripts/light_utils.py:39: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  CI_new[\"gene_strandR1B\"] = CI_new[\"refName_ABR1\"].str.split(\"|\").str[1].str.split(\":\").str[-1]\n",
      "/home/jovyan/scripts/light_utils.py:40: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  CI_new[\"numgeneR1A\"] = CI_new[\"geneR1A\"].str.split(\",\").str.len()#.astype(\"Int32\") # the number of overlapping genes on the R1A side\n",
      "/home/jovyan/scripts/light_utils.py:41: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  CI_new[[\"numgeneR1A\"]] = CI_new[[\"numgeneR1A\"]].fillna(0)\n",
      "/home/jovyan/scripts/light_utils.py:42: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  CI_new[\"numgeneR1B\"] = CI_new[\"geneR1B\"].str.split(\",\").str.len()#.astype(\"Int32\") # the number of overlapping genes on the R1B side\n",
      "/home/jovyan/scripts/light_utils.py:43: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  CI_new[[\"numgeneR1B\"]] = CI_new[[\"numgeneR1B\"]].fillna(0)\n",
      "/home/jovyan/scripts/light_utils.py:52: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  CI_new[\"shared_gene\"] = [\",\".join([x for x in a.split(\",\") if x in b.split(\",\")]) for a,b in zip(CI_new[\"geneR1A\"],CI_new[\"geneR1B\"])]\n",
      "/home/jovyan/scripts/light_utils.py:55: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  CI_new[\"num_shared_genes\"] = CI_new[\"shared_gene\"].str.split(\",\").str.len()\n",
      "/home/jovyan/scripts/light_utils.py:61: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  CI_new[\"geneR1A_uniq\"] = CI_new[\"geneR1A\"]\n",
      "/home/jovyan/scripts/light_utils.py:62: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  CI_new[\"geneR1B_uniq\"] = CI_new[\"geneR1B\"]\n",
      "/home/jovyan/scripts/light_utils.py:113: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.\n",
      "The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.\n",
      "\n",
      "For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.\n",
      "\n",
      "\n",
      "  CI_new[\"geneR1B_uniq\"].fillna(\"\",inplace=True)\n",
      "/home/jovyan/scripts/light_utils.py:114: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.\n",
      "The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.\n",
      "\n",
      "For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.\n",
      "\n",
      "\n",
      "  CI_new[\"geneR1A_uniq\"].fillna(\"\",inplace=True)\n",
      "ended modify 6.377323150634766\n",
      "total time 0:00:06.973297\n",
      "Thu Sep 26 05:35:08 PM UTC 2024 (run_class_input.sh)\n",
      "Thu Sep 26 05:35:08 PM UTC 2024\n",
      "Loading required package: data.table\n",
      "Loading required package: glmnet\n",
      "Loading required package: Matrix\n",
      "Loaded glmnet 4.1-8\n",
      "Loading required package: tictoc\n",
      "\n",
      "Attaching package: ‘tictoc’\n",
      "\n",
      "The following object is masked from ‘package:data.table’:\n",
      "\n",
      "    shift\n",
      "\n",
      "Loading required package: dplyr\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "Loading required package: stringr\n",
      "Loading required package: cutpointr\n",
      "Loading required package: GenomicAlignments\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "Loading required package: stats4\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    first, rename\n",
      "\n",
      "The following objects are masked from ‘package:Matrix’:\n",
      "\n",
      "    expand, unname\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    first, second\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Attaching package: ‘IRanges’\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    collapse, desc, slice\n",
      "\n",
      "The following object is masked from ‘package:tictoc’:\n",
      "\n",
      "    shift\n",
      "\n",
      "The following object is masked from ‘package:data.table’:\n",
      "\n",
      "    shift\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "Loading required package: GenomicRanges\n",
      "Loading required package: SummarizedExperiment\n",
      "Loading required package: MatrixGenerics\n",
      "Loading required package: matrixStats\n",
      "\n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "Loading required package: Biobase\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "Loading required package: Biostrings\n",
      "Loading required package: XVector\n",
      "\n",
      "Attaching package: ‘Biostrings’\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    strsplit\n",
      "\n",
      "Loading required package: Rsamtools\n",
      "\n",
      "Attaching package: ‘GenomicAlignments’\n",
      "\n",
      "The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    last\n",
      "\n",
      "The following object is masked from ‘package:data.table’:\n",
      "\n",
      "    last\n",
      "\n",
      "reading inputs:: 0 sec elapsed\n",
      "unstranded data modification: 3.289 sec elapsed\n",
      "Warning messages:\n",
      "1: In `[.data.table`(class_input, , `:=`(c(\"junc_cdf_glm\", \"junc_cdf_glm_corrected\",  :\n",
      "  Column 'junc_cdf_glm' does not exist to remove\n",
      "2: In `[.data.table`(class_input, , `:=`(c(\"junc_cdf_glm\", \"junc_cdf_glm_corrected\",  :\n",
      "  Column 'junc_cdf_glm_corrected' does not exist to remove\n",
      "3: In `[.data.table`(class_input, , `:=`(c(\"junc_cdf_glm\", \"junc_cdf_glm_corrected\",  :\n",
      "  Column 'junc_cdf_glmnet' does not exist to remove\n",
      "4: In `[.data.table`(class_input, , `:=`(c(\"junc_cdf_glm\", \"junc_cdf_glm_corrected\",  :\n",
      "  Column 'junc_cdf_glmnet_constrained' does not exist to remove\n",
      "5: In `[.data.table`(class_input, , `:=`(c(\"junc_cdf_glm\", \"junc_cdf_glm_corrected\",  :\n",
      "  Column 'junc_cdf_glmnet_corrected' does not exist to remove\n",
      "6: In `[.data.table`(class_input, , `:=`(c(\"junc_cdf_glm\", \"junc_cdf_glm_corrected\",  :\n",
      "  Column 'junc_cdf_glmnet_corrected_constrained' does not exist to remove\n",
      "Warning messages:\n",
      "1: In `[.data.table`(class_input, , `:=`(c(\"p_predicted_glm\", \"p_predicted_corrected\",  :\n",
      "  Column 'p_predicted_glm' does not exist to remove\n",
      "2: In `[.data.table`(class_input, , `:=`(c(\"p_predicted_glm\", \"p_predicted_corrected\",  :\n",
      "  Column 'p_predicted_corrected' does not exist to remove\n",
      "3: In `[.data.table`(class_input, , `:=`(c(\"p_predicted_glm\", \"p_predicted_corrected\",  :\n",
      "  Column 'p_predicted_glmnet' does not exist to remove\n",
      "4: In `[.data.table`(class_input, , `:=`(c(\"p_predicted_glm\", \"p_predicted_corrected\",  :\n",
      "  Column 'p_predicted_glmnet_constrained' does not exist to remove\n",
      "5: In `[.data.table`(class_input, , `:=`(c(\"p_predicted_glm\", \"p_predicted_corrected\",  :\n",
      "  Column 'p_predicted_glmnet_corrected' does not exist to remove\n",
      "6: In `[.data.table`(class_input, , `:=`(c(\"p_predicted_glm\", \"p_predicted_corrected\",  :\n",
      "  Column 'p_predicted_glmnet_corrected_constrained' does not exist to remove\n",
      "Entire code: 0.754 sec elapsed\n",
      "Warning message:\n",
      "In `[.data.table`(class_input, , `:=`(cur_weight, NULL)) :\n",
      "  Column 'cur_weight' does not exist to remove\n",
      "Warning message:\n",
      "In `[.data.table`(class_input, , `:=`(train_class, NULL)) :\n",
      "  Column 'train_class' does not exist to remove\n",
      "[1] \"GLMnet constrained\"\n",
      "[1] \"done with fitting GLMnet constrained\"\n",
      "GLMnet constrained: 1.287 sec elapsed\n",
      "13 x 1 sparse Matrix of class \"dgCMatrix\"\n",
      "                                   s1\n",
      "(Intercept)                 .        \n",
      "(Intercept)                 .        \n",
      "overlap_R1                  .        \n",
      "max_overlap_R1              .        \n",
      "NHR1A                      -1.0808247\n",
      "nmmR1                       .        \n",
      "entropyR1                   .        \n",
      "length_adj_AS_R1            2.5564956\n",
      "overlap_R1:max_overlap_R1   .        \n",
      "MR1A:SR1A                   .        \n",
      "MR1B:SR1B                   .        \n",
      "NHR1A:entropyR1             .        \n",
      "entropyR1:length_adj_AS_R1  0.1079984\n",
      "\n",
      "Call:  cv.glmnet(x = x_glmnet, y = as.factor(class_input[!is.na(train_class)]$train_class),      weights = class_input[!is.na(train_class)]$cur_weight, nfolds = 5,      family = c(\"binomial\"), intercept = FALSE, alpha = 1, nlambda = 50,      upper.limits = c(rep(Inf, 3), 0, 0, rep(Inf, 7)), lower.limits = c(-Inf,          0, 0, rep(-Inf, 2), 0, 0, rep(-Inf, 4), 0)) \n",
      "\n",
      "Measure: Binomial Deviance \n",
      "\n",
      "      Lambda Index Measure      SE Nonzero\n",
      "min 0.000155    41  0.8540 0.02386       8\n",
      "1se 0.024810    14  0.8758 0.01601       3\n",
      "[1] \"total reads: 5238\"\n",
      "[1] \"both negative 1581 out of  2106\"\n",
      "[1] \"both positive 2700 out of  3132\"\n",
      "[1] \"classification errors for glm 957 out of 5238 0.18270332187858\"\n",
      "[1] \"done with GLMnet constrained\"\n",
      "[1] \"Two-step GLMnet constrained\"\n",
      "[1] \"done with fitting Two-step GLMnet constrained\"\n",
      "9 x 1 sparse Matrix of class \"dgCMatrix\"\n",
      "                                  s1\n",
      "(Intercept)                .        \n",
      "(Intercept)                .        \n",
      "overlap_R1                 .        \n",
      "max_overlap_R1             .        \n",
      "nmmR1                     -0.3307820\n",
      "length_adj_AS_R1A          .        \n",
      "length_adj_AS_R1B          .        \n",
      "entropyR1                  0.3750278\n",
      "overlap_R1:max_overlap_R1  .        \n",
      "Two-step GLMnet constrained: 0.184 sec elapsed\n",
      "[1] \"total reads: 135\"\n",
      "[1] \"both negative 32 out of  45\"\n",
      "[1] \"both positive 71 out of  90\"\n",
      "[1] \"classification errors for glm 32 out of 135 0.237037037037037\"\n",
      "junc_median_p_val: 5.003 sec elapsed\n",
      "add_junc_seq: 0.204 sec elapsed\n",
      "Warning messages:\n",
      "1: In `[.data.table`(class_input, , `:=`(c(\"cur_weight\", \"train_class\",  :\n",
      "  Column 'junc_cdf1_glmnet_twostep' does not exist to remove\n",
      "2: In `[.data.table`(class_input, , `:=`(c(\"cur_weight\", \"train_class\",  :\n",
      "  Column 'refName_readStrandR1' does not exist to remove\n",
      "3: In `[.data.table`(class_input, , `:=`(c(\"cur_weight\", \"train_class\",  :\n",
      "  Column 'refName_readStrandR2' does not exist to remove\n",
      "4: In `[.data.table`(class_input, , `:=`(c(\"cur_weight\", \"train_class\",  :\n",
      "  Column 'gene_strandR1A_new' does not exist to remove\n",
      "5: In `[.data.table`(class_input, , `:=`(c(\"cur_weight\", \"train_class\",  :\n",
      "  Column 'gene_strandR1B_new' does not exist to remove\n",
      "7 x 1 sparse Matrix of class \"dgCMatrix\"\n",
      "                               s1\n",
      "(Intercept)                     .\n",
      "(Intercept)                     .\n",
      "discrete_overlap                .\n",
      "ave_entropyR1                   .\n",
      "ave_max_run_R1                  .\n",
      "max_njunc                       .\n",
      "discrete_overlap:ave_entropyR1  .\n",
      "Assuming the positive class is 1\n",
      "Assuming the positive class has higher x values\n",
      "Multiple optimal cutpoints found, applying break_ties.\n",
      "Warning messages:\n",
      "1: In `[.data.table`(GLM_output, , `:=`(frac_mutimapping, NULL)) :\n",
      "  Column 'frac_mutimapping' does not exist to remove\n",
      "2: In `[.data.table`(GLM_output, , `:=`(train, NULL)) :\n",
      "  Column 'train' does not exist to remove\n",
      "25.366 sec elapsed\n",
      "Thu Sep 26 05:35:46 PM UTC 2024 (run_GLM.sh)\n"
     ]
    }
   ],
   "source": [
    "%run SICILIAN.py"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d2f266d6-cf54-46f8-8f9e-66f3e05cc516",
   "metadata": {},
   "source": [
    "Note that **if you see \"Multiple optimal cutpoints found, applying break_ties\", about eight lines from the end** of the output that you should see above, then **it indicates it worked**, despite warnings like `Column 'frac_mutimapping' does not exist to remove` and `Column 'train' does not exist to remove` near the end.\n",
    "\n",
    "We can further show that by looking at some results in `sicilian_output/`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "95ff8ad3-6701-4e99-96da-ec2799d2408c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "refName_newR1\tnumReads\tgeneR1A_uniq\tgeneR1B_uniq\tchrR1A\tjuncPosR1A\tjuncPosR1B\tgene_strandR1A\tintron_length\tseqR1\tsd_overlap\tfrac_genomic_reads\tave_AT_run_R1\tave_GC_run_R1\tave_max_run_R1\tjunc_cdf_glmnet_constrained\temp.p_glmnet_constrained\tave_entropyR1\texon_annR1A\texon_annR1B\tboth_ann\tsort_junc\tsplice_ann\n",
      "chr21:AP000320.7:34125480:+|chr21:AP000320.7:34142408:+|lin\t2\tAP000320.7\tAP000320.7\tchr21\t34125480\t34142408\t+\t16928\tGGATCTCTGCCCACAGTCAGCAGCACAACAGAGGCGGGTATTTCTTGGTGGATTTTTATGCACCCACCGCAGCTGT\t4.24264068711928\t0\t5.0\t3.0\t5.0\t0.9906\t0.0\t4.2285\tTrue\tTrue\tTrue\t(34125480, 34142408)\tTrue\n",
      "chr21:ATP5J:25729631:-|chr21:ATP5J:25725350:-|lin\t6\tATP5J\tATP5J\tchr21\t25729631\t25725350\t-\t4281\tCTGATACTCTGAACTAGCATCAACAGGTCCTCCAGATGTCTGTCGCTTAGATTTGTATTCTCTAATCTTGTCCACA\t5.46808924579693\t0\t3.0\t2.0\t3.0\t1.0\t0.0\t4.2575\tTrue\tTrue\tTrue\t(25725350, 25729631)\tFalse\n",
      "chr21:ATP5J:25734853:-|chr21:ATP5J:25729801:-|lin\t5\tATP5J\tATP5J\tchr21\t25734853\t25729801\t-\t5052\tAATGCCAGAGGAGAACCTGAAGAGCCTCTGAAGAATCATGCTGATTCTTGCACTCAGTCCCGAGCTGCCAAAGCCT\t11.4673449411797\t0\t2.8\t3.2\t3.2\t0.9966\t0.0\t4.165\tTrue\tTrue\tTrue\t(25729801, 25734853)\tFalse\n",
      "chr21:ATP5O:33903935:-|chr21:ATP5O:33903639:-|lin\t8\tATP5O\tATP5O\tchr21\t33903935\t33903639\t-\t296\tATGCGCACAATCATTCCACCCAAGATTGACGGATCAGTCTTAGCCTCCAATTTCAATACTTGGCCTTGACTTAGGA\t11.7594946441467\t0\t3.5\t2.75\t3.5\t1.0\t0.0\t4.238\tTrue\tTrue\tTrue\t(33903639, 33903935)\tFalse\n",
      "chr21:ATP5O:33907341:-|chr21:ATP5O:33904021:-|lin\t9\tATP5O\tATP5O\tchr21\t33907341\t33904021\t-\t3320\tTTAATTCAGAGAGTGTGGCTTCTTCTAAAGGAGATGCAGAGGTCACTGTGCAAGGTACCTCTCCGCGATGGACACT\t7.09068246206088\t0\t4.0\t2.0\t4.0\t1.0\t0.0\t4.24444444444444\tTrue\tTrue\tTrue\t(33904021, 33907341)\tFalse\n",
      "chr21:ATP5O:33909082:-|chr21:ATP5O:33907453:-|lin\t9\tATP5O\tATP5O\tchr21\t33909082\t33907453\t-\t1629\tCCTTGGGTATTGCTTAATCGCCCATTTTCAGCAAGCAAATTGATCAGGTTGGTAGTGAGGGGAGAGAACCTCTCTT\t7.92148975887743\t0\t4.0\t3.44444444444444\t4.0\t1.0\t0.0\t4.24855555555555\tTrue\tTrue\tTrue\t(33907453, 33909082)\tFalse\n",
      "chr21:ATP5O:33912289:-|chr21:ATP5O:33909211:-|lin\t12\tATP5O\tATP5O\tchr21\t33912289\t33909211\t-\t3078\tCAGAAGCAGCCACTTTGGGTTCCTTCAGGATTTGTGCTACTCTCAACAACTCCTTTTCTACTTGCTCCAGCTTATT\t8.65850431481656\t0\t4.0\t2.75\t4.0\t1.0\t0.0\t4.23158333333333\tTrue\tTrue\tTrue\t(33909211, 33912289)\tFalse\n",
      "chr21:BRWD1:39293811:-|chr21:BRWD1:39280248:-|lin\t3\tBRWD1\tBRWD1\tchr21\t39293811\t39280248\t-\t13563\tCCATGTATCTTTGAGAGCCTTTGGCCATCGGGCTAAACTGTAAAGATGTAATTGATCCTGTGTGTCCTTGGAGCAC\t13.114877048604\t0\t3.0\t3.0\t3.0\t0.9994\t0.0\t4.19966666666667\tTrue\tTrue\tTrue\t(39280248, 39293811)\tFalse\n",
      "chr21:BRWD1:39313072:-|chr21:BRWD1:39312900:-|lin\t2\tBRWD1\tBRWD1\tchr21\t39313072\t39312900\t-\t172\tCGGCAACAACTGGTACTGCTCCAGCTCCTGCACCAGCACCTGGGCCGCTCTCCGACACGGGCCCGCCGATAGGTAC\t4.24264068711928\t0\t2.0\t3.0\t3.0\t1.0\t0.0\t4.1705\tTrue\tTrue\tTrue\t(39312900, 39313072)\tFalse\n"
     ]
    }
   ],
   "source": [
    "!head sicilian_output/human_21_hiPSC_data/demo_input/chr21_mapped_SRR12269911/sicilian_called_splice_juncs.tsv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dff054bf-bd5c-451b-aefa-f38a33b3cacd",
   "metadata": {},
   "source": [
    "Certainly some data there. \n",
    "You'd probably want to take that and use it to analyze the data in which you were interested.  \n",
    "Without doing that separately for now, let's get a better sense of the amount of information here using Pandas by running the next four cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "0f372798-d00e-4ef4-9aef-c5212a3c1e53",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "%pip install sweetviz skimpy ydata-profiling -q"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7d5cfbe5-5fd7-4e9b-9d09-3d6bee21d5a8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">╭──────────────────────────────────────────────── skimpy summary ─────────────────────────────────────────────────╮\n",
       "│ <span style=\"font-style: italic\">         Data Summary         </span> <span style=\"font-style: italic\">      Data Types       </span>                                                          │\n",
       "│ ┏━━━━━━━━━━━━━━━━━━━┳━━━━━━━━┓ ┏━━━━━━━━━━━━━┳━━━━━━━┓                                                          │\n",
       "│ ┃<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\"> dataframe         </span>┃<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\"> Values </span>┃ ┃<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\"> Column Type </span>┃<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\"> Count </span>┃                                                          │\n",
       "│ ┡━━━━━━━━━━━━━━━━━━━╇━━━━━━━━┩ ┡━━━━━━━━━━━━━╇━━━━━━━┩                                                          │\n",
       "│ │ Number of rows    │ 165    │ │ string      │ 7     │                                                          │\n",
       "│ │ Number of columns │ 23     │ │ float64     │ 7     │                                                          │\n",
       "│ └───────────────────┴────────┘ │ int64       │ 5     │                                                          │\n",
       "│                                │ bool        │ 4     │                                                          │\n",
       "│                                └─────────────┴───────┘                                                          │\n",
       "│ <span style=\"font-style: italic\">                                                    number                                                    </span>  │\n",
       "│ ┏━━━━━━━━━━━┳━━━━┳━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━┓  │\n",
       "│ ┃<span style=\"font-weight: bold\"> column_na </span>┃<span style=\"font-weight: bold\">    </span>┃<span style=\"font-weight: bold\">      </span>┃<span style=\"font-weight: bold\">          </span>┃<span style=\"font-weight: bold\">          </span>┃<span style=\"font-weight: bold\">         </span>┃<span style=\"font-weight: bold\">          </span>┃<span style=\"font-weight: bold\">          </span>┃<span style=\"font-weight: bold\">          </span>┃<span style=\"font-weight: bold\">          </span>┃<span style=\"font-weight: bold\">        </span>┃  │\n",
       "│ ┃<span style=\"font-weight: bold\"> me        </span>┃<span style=\"font-weight: bold\"> NA </span>┃<span style=\"font-weight: bold\"> NA % </span>┃<span style=\"font-weight: bold\"> mean     </span>┃<span style=\"font-weight: bold\"> sd       </span>┃<span style=\"font-weight: bold\"> p0      </span>┃<span style=\"font-weight: bold\"> p25      </span>┃<span style=\"font-weight: bold\"> p50      </span>┃<span style=\"font-weight: bold\"> p75      </span>┃<span style=\"font-weight: bold\"> p100     </span>┃<span style=\"font-weight: bold\"> hist   </span>┃  │\n",
       "│ ┡━━━━━━━━━━━╇━━━━╇━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━┩  │\n",
       "│ │ <span style=\"color: #af87ff; text-decoration-color: #af87ff\">numReads </span> │ <span style=\"color: #008080; text-decoration-color: #008080\"> 0</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">   0</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">   5.588</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">   7.213</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">      2</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">       2</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">       3</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">       5</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">      41</span> │ <span style=\"color: #008000; text-decoration-color: #008000\">  ▇▁  </span> │  │\n",
       "│ │ <span style=\"color: #af87ff; text-decoration-color: #af87ff\">juncPosR1</span> │ <span style=\"color: #008080; text-decoration-color: #008080\"> 0</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">   0</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">31610000</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">11950000</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">5126000</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">28890000</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">33600000</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">40120000</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">46660000</span> │ <span style=\"color: #008000; text-decoration-color: #008000\">▃▁▂▇▇▇</span> │  │\n",
       "│ │ <span style=\"color: #af87ff; text-decoration-color: #af87ff\">A        </span> │    │      │          │          │         │          │          │          │          │        │  │\n",
       "│ │ <span style=\"color: #af87ff; text-decoration-color: #af87ff\">juncPosR1</span> │ <span style=\"color: #008080; text-decoration-color: #008080\"> 0</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">   0</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">31620000</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">11940000</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">5128000</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">28890000</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">33600000</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">39960000</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">46660000</span> │ <span style=\"color: #008000; text-decoration-color: #008000\">▃▁▂▆▇▇</span> │  │\n",
       "│ │ <span style=\"color: #af87ff; text-decoration-color: #af87ff\">B        </span> │    │      │          │          │         │          │          │          │          │        │  │\n",
       "│ │ <span style=\"color: #af87ff; text-decoration-color: #af87ff\">intron_le</span> │ <span style=\"color: #008080; text-decoration-color: #008080\"> 0</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">   0</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">   53080</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">  160600</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">     75</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">     956</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">    2820</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">   17260</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">  991600</span> │ <span style=\"color: #008000; text-decoration-color: #008000\">  ▇   </span> │  │\n",
       "│ │ <span style=\"color: #af87ff; text-decoration-color: #af87ff\">ngth     </span> │    │      │          │          │         │          │          │          │          │        │  │\n",
       "│ │ <span style=\"color: #af87ff; text-decoration-color: #af87ff\">sd_overla</span> │ <span style=\"color: #008080; text-decoration-color: #008080\"> 0</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">   0</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">   6.933</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">   4.853</span> │ <span style=\"color: #008080; text-decoration-color: #008080\"> 0.4472</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">   2.309</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">   7.071</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">   10.07</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">   21.21</span> │ <span style=\"color: #008000; text-decoration-color: #008000\">▇▅▆▃▁ </span> │  │\n",
       "│ │ <span style=\"color: #af87ff; text-decoration-color: #af87ff\">p        </span> │    │      │          │          │         │          │          │          │          │        │  │\n",
       "│ │ <span style=\"color: #af87ff; text-decoration-color: #af87ff\">frac_geno</span> │ <span style=\"color: #008080; text-decoration-color: #008080\"> 0</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">   0</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">       0</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">       0</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">      0</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">       0</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">       0</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">       0</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">       0</span> │ <span style=\"color: #008000; text-decoration-color: #008000\">    ▇ </span> │  │\n",
       "│ │ <span style=\"color: #af87ff; text-decoration-color: #af87ff\">mic_reads</span> │    │      │          │          │         │          │          │          │          │        │  │\n",
       "│ │ <span style=\"color: #af87ff; text-decoration-color: #af87ff\">ave_AT_ru</span> │ <span style=\"color: #008080; text-decoration-color: #008080\"> 0</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">   0</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">   3.432</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">   1.226</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">  1.667</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">   2.667</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">       3</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">       4</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">   7.667</span> │ <span style=\"color: #008000; text-decoration-color: #008000\">▅▇▅▂▁▁</span> │  │\n",
       "│ │ <span style=\"color: #af87ff; text-decoration-color: #af87ff\">n_R1     </span> │    │      │          │          │         │          │          │          │          │        │  │\n",
       "│ │ <span style=\"color: #af87ff; text-decoration-color: #af87ff\">ave_GC_ru</span> │ <span style=\"color: #008080; text-decoration-color: #008080\"> 0</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">   0</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">   3.175</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">  0.7787</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">    1.5</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">       3</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">       3</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">     3.5</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">       6</span> │ <span style=\"color: #008000; text-decoration-color: #008000\"> ▁▂▇▂ </span> │  │\n",
       "│ │ <span style=\"color: #af87ff; text-decoration-color: #af87ff\">n_R1     </span> │    │      │          │          │         │          │          │          │          │        │  │\n",
       "│ │ <span style=\"color: #af87ff; text-decoration-color: #af87ff\">ave_max_r</span> │ <span style=\"color: #008080; text-decoration-color: #008080\"> 0</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">   0</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">   3.923</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">   1.022</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">  2.333</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">       3</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">   3.783</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">       4</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">   7.667</span> │ <span style=\"color: #008000; text-decoration-color: #008000\">▆▇▁▁▁▁</span> │  │\n",
       "│ │ <span style=\"color: #af87ff; text-decoration-color: #af87ff\">un_R1    </span> │    │      │          │          │         │          │          │          │          │        │  │\n",
       "│ │ <span style=\"color: #af87ff; text-decoration-color: #af87ff\">junc_cdf_</span> │ <span style=\"color: #008080; text-decoration-color: #008080\"> 0</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">   0</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">  0.8918</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">  0.1555</span> │ <span style=\"color: #008080; text-decoration-color: #008080\"> 0.4186</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">  0.8091</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">  0.9821</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">  0.9999</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">       1</span> │ <span style=\"color: #008000; text-decoration-color: #008000\">  ▁▁▁▇</span> │  │\n",
       "│ │ <span style=\"color: #af87ff; text-decoration-color: #af87ff\">glmnet_co</span> │    │      │          │          │         │          │          │          │          │        │  │\n",
       "│ │ <span style=\"color: #af87ff; text-decoration-color: #af87ff\">nstrained</span> │    │      │          │          │         │          │          │          │          │        │  │\n",
       "│ │ <span style=\"color: #af87ff; text-decoration-color: #af87ff\">emp.p_glm</span> │ <span style=\"color: #008080; text-decoration-color: #008080\"> 0</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">   0</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">0.004398</span> │ <span style=\"color: #008080; text-decoration-color: #008080\"> 0.01077</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">      0</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">       0</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">       0</span> │ <span style=\"color: #008080; text-decoration-color: #008080\"> 0.00509</span> │ <span style=\"color: #008080; text-decoration-color: #008080\"> 0.08767</span> │ <span style=\"color: #008000; text-decoration-color: #008000\">  ▇   </span> │  │\n",
       "│ │ <span style=\"color: #af87ff; text-decoration-color: #af87ff\">net_const</span> │    │      │          │          │         │          │          │          │          │        │  │\n",
       "│ │ <span style=\"color: #af87ff; text-decoration-color: #af87ff\">rained   </span> │    │      │          │          │         │          │          │          │          │        │  │\n",
       "│ │ <span style=\"color: #af87ff; text-decoration-color: #af87ff\">ave_entro</span> │ <span style=\"color: #008080; text-decoration-color: #008080\"> 0</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">   0</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">   4.192</span> │ <span style=\"color: #008080; text-decoration-color: #008080\"> 0.05711</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">  3.854</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">   4.168</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">     4.2</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">   4.231</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">   4.277</span> │ <span style=\"color: #008000; text-decoration-color: #008000\">   ▂▇▇</span> │  │\n",
       "│ │ <span style=\"color: #af87ff; text-decoration-color: #af87ff\">pyR1     </span> │    │      │          │          │         │          │          │          │          │        │  │\n",
       "│ └───────────┴────┴──────┴──────────┴──────────┴─────────┴──────────┴──────────┴──────────┴──────────┴────────┘  │\n",
       "│ <span style=\"font-style: italic\">                                                     bool                                                     </span>  │\n",
       "│ ┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━┓  │\n",
       "│ ┃<span style=\"font-weight: bold\"> column_name                        </span>┃<span style=\"font-weight: bold\"> true            </span>┃<span style=\"font-weight: bold\"> true rate                     </span>┃<span style=\"font-weight: bold\"> hist                </span>┃  │\n",
       "│ ┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━┩  │\n",
       "│ │ <span style=\"color: #af87ff; text-decoration-color: #af87ff\">exon_annR1A                       </span> │ <span style=\"color: #008080; text-decoration-color: #008080\">            108</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">                         0.65</span> │ <span style=\"color: #008000; text-decoration-color: #008000\">      ▅    ▇       </span> │  │\n",
       "│ │ <span style=\"color: #af87ff; text-decoration-color: #af87ff\">exon_annR1B                       </span> │ <span style=\"color: #008080; text-decoration-color: #008080\">            107</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">                         0.65</span> │ <span style=\"color: #008000; text-decoration-color: #008000\">      ▅    ▇       </span> │  │\n",
       "│ │ <span style=\"color: #af87ff; text-decoration-color: #af87ff\">both_ann                          </span> │ <span style=\"color: #008080; text-decoration-color: #008080\">            107</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">                         0.65</span> │ <span style=\"color: #008000; text-decoration-color: #008000\">      ▅    ▇       </span> │  │\n",
       "│ │ <span style=\"color: #af87ff; text-decoration-color: #af87ff\">splice_ann                        </span> │ <span style=\"color: #008080; text-decoration-color: #008080\">             34</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">                         0.21</span> │ <span style=\"color: #008000; text-decoration-color: #008000\">      ▇    ▂       </span> │  │\n",
       "│ └────────────────────────────────────┴─────────────────┴───────────────────────────────┴─────────────────────┘  │\n",
       "│ <span style=\"font-style: italic\">                                                    string                                                    </span>  │\n",
       "│ ┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┓  │\n",
       "│ ┃<span style=\"font-weight: bold\"> column_name                   </span>┃<span style=\"font-weight: bold\"> NA     </span>┃<span style=\"font-weight: bold\"> NA %       </span>┃<span style=\"font-weight: bold\"> words per row               </span>┃<span style=\"font-weight: bold\"> total words            </span>┃  │\n",
       "│ ┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━┩  │\n",
       "│ │ <span style=\"color: #af87ff; text-decoration-color: #af87ff\">refName_newR1                </span> │ <span style=\"color: #008080; text-decoration-color: #008080\">     0</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">         0</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">                          1</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">                   165</span> │  │\n",
       "│ │ <span style=\"color: #af87ff; text-decoration-color: #af87ff\">geneR1A_uniq                 </span> │ <span style=\"color: #008080; text-decoration-color: #008080\">     0</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">         0</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">                          1</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">                   165</span> │  │\n",
       "│ │ <span style=\"color: #af87ff; text-decoration-color: #af87ff\">geneR1B_uniq                 </span> │ <span style=\"color: #008080; text-decoration-color: #008080\">     0</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">         0</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">                          1</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">                   165</span> │  │\n",
       "│ │ <span style=\"color: #af87ff; text-decoration-color: #af87ff\">chrR1A                       </span> │ <span style=\"color: #008080; text-decoration-color: #008080\">     0</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">         0</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">                          1</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">                   165</span> │  │\n",
       "│ │ <span style=\"color: #af87ff; text-decoration-color: #af87ff\">gene_strandR1A               </span> │ <span style=\"color: #008080; text-decoration-color: #008080\">     0</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">         0</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">                          1</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">                   165</span> │  │\n",
       "│ │ <span style=\"color: #af87ff; text-decoration-color: #af87ff\">seqR1                        </span> │ <span style=\"color: #008080; text-decoration-color: #008080\">     0</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">         0</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">                          1</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">                   165</span> │  │\n",
       "│ │ <span style=\"color: #af87ff; text-decoration-color: #af87ff\">sort_junc                    </span> │ <span style=\"color: #008080; text-decoration-color: #008080\">     0</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">         0</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">                          2</span> │ <span style=\"color: #008080; text-decoration-color: #008080\">                   330</span> │  │\n",
       "│ └───────────────────────────────┴────────┴────────────┴─────────────────────────────┴────────────────────────┘  │\n",
       "╰────────────────────────────────────────────────────── End ──────────────────────────────────────────────────────╯\n",
       "</pre>\n"
      ],
      "text/plain": [
       "╭──────────────────────────────────────────────── skimpy summary ─────────────────────────────────────────────────╮\n",
       "│ \u001b[3m         Data Summary         \u001b[0m \u001b[3m      Data Types       \u001b[0m                                                          │\n",
       "│ ┏━━━━━━━━━━━━━━━━━━━┳━━━━━━━━┓ ┏━━━━━━━━━━━━━┳━━━━━━━┓                                                          │\n",
       "│ ┃\u001b[1;36m \u001b[0m\u001b[1;36mdataframe        \u001b[0m\u001b[1;36m \u001b[0m┃\u001b[1;36m \u001b[0m\u001b[1;36mValues\u001b[0m\u001b[1;36m \u001b[0m┃ ┃\u001b[1;36m \u001b[0m\u001b[1;36mColumn Type\u001b[0m\u001b[1;36m \u001b[0m┃\u001b[1;36m \u001b[0m\u001b[1;36mCount\u001b[0m\u001b[1;36m \u001b[0m┃                                                          │\n",
       "│ ┡━━━━━━━━━━━━━━━━━━━╇━━━━━━━━┩ ┡━━━━━━━━━━━━━╇━━━━━━━┩                                                          │\n",
       "│ │ Number of rows    │ 165    │ │ string      │ 7     │                                                          │\n",
       "│ │ Number of columns │ 23     │ │ float64     │ 7     │                                                          │\n",
       "│ └───────────────────┴────────┘ │ int64       │ 5     │                                                          │\n",
       "│                                │ bool        │ 4     │                                                          │\n",
       "│                                └─────────────┴───────┘                                                          │\n",
       "│ \u001b[3m                                                    number                                                    \u001b[0m  │\n",
       "│ ┏━━━━━━━━━━━┳━━━━┳━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━┓  │\n",
       "│ ┃\u001b[1m \u001b[0m\u001b[1mcolumn_na\u001b[0m\u001b[1m \u001b[0m┃\u001b[1m    \u001b[0m┃\u001b[1m      \u001b[0m┃\u001b[1m          \u001b[0m┃\u001b[1m          \u001b[0m┃\u001b[1m         \u001b[0m┃\u001b[1m          \u001b[0m┃\u001b[1m          \u001b[0m┃\u001b[1m          \u001b[0m┃\u001b[1m          \u001b[0m┃\u001b[1m        \u001b[0m┃  │\n",
       "│ ┃\u001b[1m \u001b[0m\u001b[1mme       \u001b[0m\u001b[1m \u001b[0m┃\u001b[1m \u001b[0m\u001b[1mNA\u001b[0m\u001b[1m \u001b[0m┃\u001b[1m \u001b[0m\u001b[1mNA %\u001b[0m\u001b[1m \u001b[0m┃\u001b[1m \u001b[0m\u001b[1mmean    \u001b[0m\u001b[1m \u001b[0m┃\u001b[1m \u001b[0m\u001b[1msd      \u001b[0m\u001b[1m \u001b[0m┃\u001b[1m \u001b[0m\u001b[1mp0     \u001b[0m\u001b[1m \u001b[0m┃\u001b[1m \u001b[0m\u001b[1mp25     \u001b[0m\u001b[1m \u001b[0m┃\u001b[1m \u001b[0m\u001b[1mp50     \u001b[0m\u001b[1m \u001b[0m┃\u001b[1m \u001b[0m\u001b[1mp75     \u001b[0m\u001b[1m \u001b[0m┃\u001b[1m \u001b[0m\u001b[1mp100    \u001b[0m\u001b[1m \u001b[0m┃\u001b[1m \u001b[0m\u001b[1mhist  \u001b[0m\u001b[1m \u001b[0m┃  │\n",
       "│ ┡━━━━━━━━━━━╇━━━━╇━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━┩  │\n",
       "│ │ \u001b[38;5;141mnumReads \u001b[0m │ \u001b[36m 0\u001b[0m │ \u001b[36m   0\u001b[0m │ \u001b[36m   5.588\u001b[0m │ \u001b[36m   7.213\u001b[0m │ \u001b[36m      2\u001b[0m │ \u001b[36m       2\u001b[0m │ \u001b[36m       3\u001b[0m │ \u001b[36m       5\u001b[0m │ \u001b[36m      41\u001b[0m │ \u001b[32m  ▇▁  \u001b[0m │  │\n",
       "│ │ \u001b[38;5;141mjuncPosR1\u001b[0m │ \u001b[36m 0\u001b[0m │ \u001b[36m   0\u001b[0m │ \u001b[36m31610000\u001b[0m │ \u001b[36m11950000\u001b[0m │ \u001b[36m5126000\u001b[0m │ \u001b[36m28890000\u001b[0m │ \u001b[36m33600000\u001b[0m │ \u001b[36m40120000\u001b[0m │ \u001b[36m46660000\u001b[0m │ \u001b[32m▃▁▂▇▇▇\u001b[0m │  │\n",
       "│ │ \u001b[38;5;141mA        \u001b[0m │    │      │          │          │         │          │          │          │          │        │  │\n",
       "│ │ \u001b[38;5;141mjuncPosR1\u001b[0m │ \u001b[36m 0\u001b[0m │ \u001b[36m   0\u001b[0m │ \u001b[36m31620000\u001b[0m │ \u001b[36m11940000\u001b[0m │ \u001b[36m5128000\u001b[0m │ \u001b[36m28890000\u001b[0m │ \u001b[36m33600000\u001b[0m │ \u001b[36m39960000\u001b[0m │ \u001b[36m46660000\u001b[0m │ \u001b[32m▃▁▂▆▇▇\u001b[0m │  │\n",
       "│ │ \u001b[38;5;141mB        \u001b[0m │    │      │          │          │         │          │          │          │          │        │  │\n",
       "│ │ \u001b[38;5;141mintron_le\u001b[0m │ \u001b[36m 0\u001b[0m │ \u001b[36m   0\u001b[0m │ \u001b[36m   53080\u001b[0m │ \u001b[36m  160600\u001b[0m │ \u001b[36m     75\u001b[0m │ \u001b[36m     956\u001b[0m │ \u001b[36m    2820\u001b[0m │ \u001b[36m   17260\u001b[0m │ \u001b[36m  991600\u001b[0m │ \u001b[32m  ▇   \u001b[0m │  │\n",
       "│ │ \u001b[38;5;141mngth     \u001b[0m │    │      │          │          │         │          │          │          │          │        │  │\n",
       "│ │ \u001b[38;5;141msd_overla\u001b[0m │ \u001b[36m 0\u001b[0m │ \u001b[36m   0\u001b[0m │ \u001b[36m   6.933\u001b[0m │ \u001b[36m   4.853\u001b[0m │ \u001b[36m 0.4472\u001b[0m │ \u001b[36m   2.309\u001b[0m │ \u001b[36m   7.071\u001b[0m │ \u001b[36m   10.07\u001b[0m │ \u001b[36m   21.21\u001b[0m │ \u001b[32m▇▅▆▃▁ \u001b[0m │  │\n",
       "│ │ \u001b[38;5;141mp        \u001b[0m │    │      │          │          │         │          │          │          │          │        │  │\n",
       "│ │ \u001b[38;5;141mfrac_geno\u001b[0m │ \u001b[36m 0\u001b[0m │ \u001b[36m   0\u001b[0m │ \u001b[36m       0\u001b[0m │ \u001b[36m       0\u001b[0m │ \u001b[36m      0\u001b[0m │ \u001b[36m       0\u001b[0m │ \u001b[36m       0\u001b[0m │ \u001b[36m       0\u001b[0m │ \u001b[36m       0\u001b[0m │ \u001b[32m    ▇ \u001b[0m │  │\n",
       "│ │ \u001b[38;5;141mmic_reads\u001b[0m │    │      │          │          │         │          │          │          │          │        │  │\n",
       "│ │ \u001b[38;5;141mave_AT_ru\u001b[0m │ \u001b[36m 0\u001b[0m │ \u001b[36m   0\u001b[0m │ \u001b[36m   3.432\u001b[0m │ \u001b[36m   1.226\u001b[0m │ \u001b[36m  1.667\u001b[0m │ \u001b[36m   2.667\u001b[0m │ \u001b[36m       3\u001b[0m │ \u001b[36m       4\u001b[0m │ \u001b[36m   7.667\u001b[0m │ \u001b[32m▅▇▅▂▁▁\u001b[0m │  │\n",
       "│ │ \u001b[38;5;141mn_R1     \u001b[0m │    │      │          │          │         │          │          │          │          │        │  │\n",
       "│ │ \u001b[38;5;141mave_GC_ru\u001b[0m │ \u001b[36m 0\u001b[0m │ \u001b[36m   0\u001b[0m │ \u001b[36m   3.175\u001b[0m │ \u001b[36m  0.7787\u001b[0m │ \u001b[36m    1.5\u001b[0m │ \u001b[36m       3\u001b[0m │ \u001b[36m       3\u001b[0m │ \u001b[36m     3.5\u001b[0m │ \u001b[36m       6\u001b[0m │ \u001b[32m ▁▂▇▂ \u001b[0m │  │\n",
       "│ │ \u001b[38;5;141mn_R1     \u001b[0m │    │      │          │          │         │          │          │          │          │        │  │\n",
       "│ │ \u001b[38;5;141mave_max_r\u001b[0m │ \u001b[36m 0\u001b[0m │ \u001b[36m   0\u001b[0m │ \u001b[36m   3.923\u001b[0m │ \u001b[36m   1.022\u001b[0m │ \u001b[36m  2.333\u001b[0m │ \u001b[36m       3\u001b[0m │ \u001b[36m   3.783\u001b[0m │ \u001b[36m       4\u001b[0m │ \u001b[36m   7.667\u001b[0m │ \u001b[32m▆▇▁▁▁▁\u001b[0m │  │\n",
       "│ │ \u001b[38;5;141mun_R1    \u001b[0m │    │      │          │          │         │          │          │          │          │        │  │\n",
       "│ │ \u001b[38;5;141mjunc_cdf_\u001b[0m │ \u001b[36m 0\u001b[0m │ \u001b[36m   0\u001b[0m │ \u001b[36m  0.8918\u001b[0m │ \u001b[36m  0.1555\u001b[0m │ \u001b[36m 0.4186\u001b[0m │ \u001b[36m  0.8091\u001b[0m │ \u001b[36m  0.9821\u001b[0m │ \u001b[36m  0.9999\u001b[0m │ \u001b[36m       1\u001b[0m │ \u001b[32m  ▁▁▁▇\u001b[0m │  │\n",
       "│ │ \u001b[38;5;141mglmnet_co\u001b[0m │    │      │          │          │         │          │          │          │          │        │  │\n",
       "│ │ \u001b[38;5;141mnstrained\u001b[0m │    │      │          │          │         │          │          │          │          │        │  │\n",
       "│ │ \u001b[38;5;141memp.p_glm\u001b[0m │ \u001b[36m 0\u001b[0m │ \u001b[36m   0\u001b[0m │ \u001b[36m0.004398\u001b[0m │ \u001b[36m 0.01077\u001b[0m │ \u001b[36m      0\u001b[0m │ \u001b[36m       0\u001b[0m │ \u001b[36m       0\u001b[0m │ \u001b[36m 0.00509\u001b[0m │ \u001b[36m 0.08767\u001b[0m │ \u001b[32m  ▇   \u001b[0m │  │\n",
       "│ │ \u001b[38;5;141mnet_const\u001b[0m │    │      │          │          │         │          │          │          │          │        │  │\n",
       "│ │ \u001b[38;5;141mrained   \u001b[0m │    │      │          │          │         │          │          │          │          │        │  │\n",
       "│ │ \u001b[38;5;141mave_entro\u001b[0m │ \u001b[36m 0\u001b[0m │ \u001b[36m   0\u001b[0m │ \u001b[36m   4.192\u001b[0m │ \u001b[36m 0.05711\u001b[0m │ \u001b[36m  3.854\u001b[0m │ \u001b[36m   4.168\u001b[0m │ \u001b[36m     4.2\u001b[0m │ \u001b[36m   4.231\u001b[0m │ \u001b[36m   4.277\u001b[0m │ \u001b[32m   ▂▇▇\u001b[0m │  │\n",
       "│ │ \u001b[38;5;141mpyR1     \u001b[0m │    │      │          │          │         │          │          │          │          │        │  │\n",
       "│ └───────────┴────┴──────┴──────────┴──────────┴─────────┴──────────┴──────────┴──────────┴──────────┴────────┘  │\n",
       "│ \u001b[3m                                                     bool                                                     \u001b[0m  │\n",
       "│ ┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━┓  │\n",
       "│ ┃\u001b[1m \u001b[0m\u001b[1mcolumn_name                       \u001b[0m\u001b[1m \u001b[0m┃\u001b[1m \u001b[0m\u001b[1mtrue           \u001b[0m\u001b[1m \u001b[0m┃\u001b[1m \u001b[0m\u001b[1mtrue rate                    \u001b[0m\u001b[1m \u001b[0m┃\u001b[1m \u001b[0m\u001b[1mhist               \u001b[0m\u001b[1m \u001b[0m┃  │\n",
       "│ ┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━┩  │\n",
       "│ │ \u001b[38;5;141mexon_annR1A                       \u001b[0m │ \u001b[36m            108\u001b[0m │ \u001b[36m                         0.65\u001b[0m │ \u001b[32m      ▅    ▇       \u001b[0m │  │\n",
       "│ │ \u001b[38;5;141mexon_annR1B                       \u001b[0m │ \u001b[36m            107\u001b[0m │ \u001b[36m                         0.65\u001b[0m │ \u001b[32m      ▅    ▇       \u001b[0m │  │\n",
       "│ │ \u001b[38;5;141mboth_ann                          \u001b[0m │ \u001b[36m            107\u001b[0m │ \u001b[36m                         0.65\u001b[0m │ \u001b[32m      ▅    ▇       \u001b[0m │  │\n",
       "│ │ \u001b[38;5;141msplice_ann                        \u001b[0m │ \u001b[36m             34\u001b[0m │ \u001b[36m                         0.21\u001b[0m │ \u001b[32m      ▇    ▂       \u001b[0m │  │\n",
       "│ └────────────────────────────────────┴─────────────────┴───────────────────────────────┴─────────────────────┘  │\n",
       "│ \u001b[3m                                                    string                                                    \u001b[0m  │\n",
       "│ ┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┓  │\n",
       "│ ┃\u001b[1m \u001b[0m\u001b[1mcolumn_name                  \u001b[0m\u001b[1m \u001b[0m┃\u001b[1m \u001b[0m\u001b[1mNA    \u001b[0m\u001b[1m \u001b[0m┃\u001b[1m \u001b[0m\u001b[1mNA %      \u001b[0m\u001b[1m \u001b[0m┃\u001b[1m \u001b[0m\u001b[1mwords per row              \u001b[0m\u001b[1m \u001b[0m┃\u001b[1m \u001b[0m\u001b[1mtotal words           \u001b[0m\u001b[1m \u001b[0m┃  │\n",
       "│ ┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━┩  │\n",
       "│ │ \u001b[38;5;141mrefName_newR1                \u001b[0m │ \u001b[36m     0\u001b[0m │ \u001b[36m         0\u001b[0m │ \u001b[36m                          1\u001b[0m │ \u001b[36m                   165\u001b[0m │  │\n",
       "│ │ \u001b[38;5;141mgeneR1A_uniq                 \u001b[0m │ \u001b[36m     0\u001b[0m │ \u001b[36m         0\u001b[0m │ \u001b[36m                          1\u001b[0m │ \u001b[36m                   165\u001b[0m │  │\n",
       "│ │ \u001b[38;5;141mgeneR1B_uniq                 \u001b[0m │ \u001b[36m     0\u001b[0m │ \u001b[36m         0\u001b[0m │ \u001b[36m                          1\u001b[0m │ \u001b[36m                   165\u001b[0m │  │\n",
       "│ │ \u001b[38;5;141mchrR1A                       \u001b[0m │ \u001b[36m     0\u001b[0m │ \u001b[36m         0\u001b[0m │ \u001b[36m                          1\u001b[0m │ \u001b[36m                   165\u001b[0m │  │\n",
       "│ │ \u001b[38;5;141mgene_strandR1A               \u001b[0m │ \u001b[36m     0\u001b[0m │ \u001b[36m         0\u001b[0m │ \u001b[36m                          1\u001b[0m │ \u001b[36m                   165\u001b[0m │  │\n",
       "│ │ \u001b[38;5;141mseqR1                        \u001b[0m │ \u001b[36m     0\u001b[0m │ \u001b[36m         0\u001b[0m │ \u001b[36m                          1\u001b[0m │ \u001b[36m                   165\u001b[0m │  │\n",
       "│ │ \u001b[38;5;141msort_junc                    \u001b[0m │ \u001b[36m     0\u001b[0m │ \u001b[36m         0\u001b[0m │ \u001b[36m                          2\u001b[0m │ \u001b[36m                   330\u001b[0m │  │\n",
       "│ └───────────────────────────────┴────────┴────────────┴─────────────────────────────┴────────────────────────┘  │\n",
       "╰────────────────────────────────────────────────────── End ──────────────────────────────────────────────────────╯\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "data_file = \"sicilian_output/human_21_hiPSC_data/demo_input/chr21_mapped_SRR12269911/sicilian_called_splice_juncs.tsv\"\n",
    "df = df = pd.read_csv(data_file, sep='\\t')\n",
    "from skimpy import skim\n",
    "skim(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0f9b9136-a398-4914-9b09-d4106f1e2896",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>numReads</th>\n",
       "      <th>juncPosR1A</th>\n",
       "      <th>juncPosR1B</th>\n",
       "      <th>intron_length</th>\n",
       "      <th>sd_overlap</th>\n",
       "      <th>frac_genomic_reads</th>\n",
       "      <th>ave_AT_run_R1</th>\n",
       "      <th>ave_GC_run_R1</th>\n",
       "      <th>ave_max_run_R1</th>\n",
       "      <th>junc_cdf_glmnet_constrained</th>\n",
       "      <th>emp.p_glmnet_constrained</th>\n",
       "      <th>ave_entropyR1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>165.000000</td>\n",
       "      <td>1.650000e+02</td>\n",
       "      <td>1.650000e+02</td>\n",
       "      <td>165.000000</td>\n",
       "      <td>165.000000</td>\n",
       "      <td>165.0</td>\n",
       "      <td>165.000000</td>\n",
       "      <td>165.000000</td>\n",
       "      <td>165.000000</td>\n",
       "      <td>165.000000</td>\n",
       "      <td>165.000000</td>\n",
       "      <td>165.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>5.587879</td>\n",
       "      <td>3.161318e+07</td>\n",
       "      <td>3.161815e+07</td>\n",
       "      <td>53076.654545</td>\n",
       "      <td>6.932511</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.431818</td>\n",
       "      <td>3.175386</td>\n",
       "      <td>3.923228</td>\n",
       "      <td>0.891817</td>\n",
       "      <td>0.004398</td>\n",
       "      <td>4.191913</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>7.212783</td>\n",
       "      <td>1.194572e+07</td>\n",
       "      <td>1.194069e+07</td>\n",
       "      <td>160567.552151</td>\n",
       "      <td>4.853326</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.225668</td>\n",
       "      <td>0.778719</td>\n",
       "      <td>1.022018</td>\n",
       "      <td>0.155453</td>\n",
       "      <td>0.010765</td>\n",
       "      <td>0.057114</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>2.000000</td>\n",
       "      <td>5.125992e+06</td>\n",
       "      <td>5.127898e+06</td>\n",
       "      <td>75.000000</td>\n",
       "      <td>0.447214</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.666667</td>\n",
       "      <td>1.500000</td>\n",
       "      <td>2.333333</td>\n",
       "      <td>0.418600</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.854000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.888831e+07</td>\n",
       "      <td>2.888863e+07</td>\n",
       "      <td>956.000000</td>\n",
       "      <td>2.309401</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.666667</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>0.809100</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.167538</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>3.000000</td>\n",
       "      <td>3.359561e+07</td>\n",
       "      <td>3.359573e+07</td>\n",
       "      <td>2820.000000</td>\n",
       "      <td>7.071068</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>3.782609</td>\n",
       "      <td>0.982100</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.200000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>5.000000</td>\n",
       "      <td>4.011550e+07</td>\n",
       "      <td>3.995995e+07</td>\n",
       "      <td>17262.000000</td>\n",
       "      <td>10.066446</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.000000</td>\n",
       "      <td>3.500000</td>\n",
       "      <td>4.000000</td>\n",
       "      <td>0.999900</td>\n",
       "      <td>0.005090</td>\n",
       "      <td>4.230600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>41.000000</td>\n",
       "      <td>4.666355e+07</td>\n",
       "      <td>4.666430e+07</td>\n",
       "      <td>991642.000000</td>\n",
       "      <td>21.213203</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.666667</td>\n",
       "      <td>6.000000</td>\n",
       "      <td>7.666667</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.087670</td>\n",
       "      <td>4.277000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         numReads    juncPosR1A    juncPosR1B  intron_length  sd_overlap  \\\n",
       "count  165.000000  1.650000e+02  1.650000e+02     165.000000  165.000000   \n",
       "mean     5.587879  3.161318e+07  3.161815e+07   53076.654545    6.932511   \n",
       "std      7.212783  1.194572e+07  1.194069e+07  160567.552151    4.853326   \n",
       "min      2.000000  5.125992e+06  5.127898e+06      75.000000    0.447214   \n",
       "25%      2.000000  2.888831e+07  2.888863e+07     956.000000    2.309401   \n",
       "50%      3.000000  3.359561e+07  3.359573e+07    2820.000000    7.071068   \n",
       "75%      5.000000  4.011550e+07  3.995995e+07   17262.000000   10.066446   \n",
       "max     41.000000  4.666355e+07  4.666430e+07  991642.000000   21.213203   \n",
       "\n",
       "       frac_genomic_reads  ave_AT_run_R1  ave_GC_run_R1  ave_max_run_R1  \\\n",
       "count               165.0     165.000000     165.000000      165.000000   \n",
       "mean                  0.0       3.431818       3.175386        3.923228   \n",
       "std                   0.0       1.225668       0.778719        1.022018   \n",
       "min                   0.0       1.666667       1.500000        2.333333   \n",
       "25%                   0.0       2.666667       3.000000        3.000000   \n",
       "50%                   0.0       3.000000       3.000000        3.782609   \n",
       "75%                   0.0       4.000000       3.500000        4.000000   \n",
       "max                   0.0       7.666667       6.000000        7.666667   \n",
       "\n",
       "       junc_cdf_glmnet_constrained  emp.p_glmnet_constrained  ave_entropyR1  \n",
       "count                   165.000000                165.000000     165.000000  \n",
       "mean                      0.891817                  0.004398       4.191913  \n",
       "std                       0.155453                  0.010765       0.057114  \n",
       "min                       0.418600                  0.000000       3.854000  \n",
       "25%                       0.809100                  0.000000       4.167538  \n",
       "50%                       0.982100                  0.000000       4.200000  \n",
       "75%                       0.999900                  0.005090       4.230600  \n",
       "max                       1.000000                  0.087670       4.277000  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "200e9380-93ca-4674-932a-7d6f491e9bad",
   "metadata": {},
   "outputs": [],
   "source": [
    "from ydata_profiling import ProfileReport\n",
    "profile = ProfileReport(df, title=\"Report: sicilian_called_splice_juncs\")\n",
    "profile"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a9519719-5078-4443-9c76-1e235f25567c",
   "metadata": {},
   "source": [
    "All of those data summary/description approaches are general to any Pandas dataframe, and so not really giving you much insight beyond there is a fair aomount of data to parse: 165 splice junctions SICILIAN deems high confidence with 23 columns related to each."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea31121b-e012-4fe6-a431-4246bddc0333",
   "metadata": {},
   "source": [
    "You'd probably want to take that tab-separated table and use it to analyze the data to focus on expression of genes of interest to you.  \n",
    "To give a sense here, let's focus on one gene CCT8 has a number of introns, that can easily be seen in [the Ensembl region in detail for CCT8](https://useast.ensembl.org/Homo_sapiens/Location/View?db=core;g=ENSG00000156261;r=21:29055802-29073805), porbably easiest summarized looking at the MANE Select Transcripts track for CCT8 (probably the 7th track down)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "654aded4-3da3-4d20-872c-877e747cd60e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>refName_newR1</th>\n",
       "      <th>numReads</th>\n",
       "      <th>geneR1A_uniq</th>\n",
       "      <th>geneR1B_uniq</th>\n",
       "      <th>chrR1A</th>\n",
       "      <th>juncPosR1A</th>\n",
       "      <th>juncPosR1B</th>\n",
       "      <th>gene_strandR1A</th>\n",
       "      <th>intron_length</th>\n",
       "      <th>seqR1</th>\n",
       "      <th>...</th>\n",
       "      <th>ave_GC_run_R1</th>\n",
       "      <th>ave_max_run_R1</th>\n",
       "      <th>junc_cdf_glmnet_constrained</th>\n",
       "      <th>emp.p_glmnet_constrained</th>\n",
       "      <th>ave_entropyR1</th>\n",
       "      <th>exon_annR1A</th>\n",
       "      <th>exon_annR1B</th>\n",
       "      <th>both_ann</th>\n",
       "      <th>sort_junc</th>\n",
       "      <th>splice_ann</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>chr21:CCT8:29060541:-|chr21:CCT8:29056552:-|lin</td>\n",
       "      <td>25</td>\n",
       "      <td>CCT8</td>\n",
       "      <td>CCT8</td>\n",
       "      <td>chr21</td>\n",
       "      <td>29060541</td>\n",
       "      <td>29056552</td>\n",
       "      <td>-</td>\n",
       "      <td>3989</td>\n",
       "      <td>GGAGGCTTGGGCCCACCAGCTGGTTTTGCCATGATGATCTGATCCA...</td>\n",
       "      <td>...</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>4.000000</td>\n",
       "      <td>0.999999</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.197200</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>(29056552, 29060541)</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>chr21:CCT8:29061253:-|chr21:CCT8:29060660:-|lin</td>\n",
       "      <td>34</td>\n",
       "      <td>CCT8</td>\n",
       "      <td>CCT8</td>\n",
       "      <td>chr21</td>\n",
       "      <td>29061253</td>\n",
       "      <td>29060660</td>\n",
       "      <td>-</td>\n",
       "      <td>593</td>\n",
       "      <td>CAGCTTCCAGCATGTCCTTTACAGCAGGGACTTCAGCCTCAATATC...</td>\n",
       "      <td>...</td>\n",
       "      <td>2.823529</td>\n",
       "      <td>4.647059</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.250088</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>(29060660, 29061253)</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>chr21:CCT8:29061496:-|chr21:CCT8:29061417:-|lin</td>\n",
       "      <td>24</td>\n",
       "      <td>CCT8</td>\n",
       "      <td>CCT8</td>\n",
       "      <td>chr21</td>\n",
       "      <td>29061496</td>\n",
       "      <td>29061417</td>\n",
       "      <td>-</td>\n",
       "      <td>79</td>\n",
       "      <td>TCTTAATAGCATACTGTTCAAGTCCAGGACATGTCTCTCCATATGA...</td>\n",
       "      <td>...</td>\n",
       "      <td>2.833333</td>\n",
       "      <td>3.250000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.223083</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>(29061417, 29061496)</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>chr21:CCT8:29062128:-|chr21:CCT8:29061567:-|lin</td>\n",
       "      <td>33</td>\n",
       "      <td>CCT8</td>\n",
       "      <td>CCT8</td>\n",
       "      <td>chr21</td>\n",
       "      <td>29062128</td>\n",
       "      <td>29061567</td>\n",
       "      <td>-</td>\n",
       "      <td>561</td>\n",
       "      <td>TTTCTGTTGCTCCACCTCCGGGTACAAGACGTTTATCCCTTGTAAG...</td>\n",
       "      <td>...</td>\n",
       "      <td>3.060606</td>\n",
       "      <td>3.060606</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.249121</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>(29061567, 29062128)</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>chr21:CCT8:29063352:-|chr21:CCT8:29062556:-|lin</td>\n",
       "      <td>23</td>\n",
       "      <td>CCT8</td>\n",
       "      <td>CCT8</td>\n",
       "      <td>chr21</td>\n",
       "      <td>29063352</td>\n",
       "      <td>29062556</td>\n",
       "      <td>-</td>\n",
       "      <td>796</td>\n",
       "      <td>AGTTTTACAAAGTCTTCGGAGATCCCATTTTGAGTTTAGCCTCACT...</td>\n",
       "      <td>...</td>\n",
       "      <td>2.695652</td>\n",
       "      <td>3.782609</td>\n",
       "      <td>0.999998</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.229609</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>(29062556, 29063352)</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>chr21:CCT8:29066891:-|chr21:CCT8:29066777:-|lin</td>\n",
       "      <td>23</td>\n",
       "      <td>CCT8</td>\n",
       "      <td>CCT8</td>\n",
       "      <td>chr21</td>\n",
       "      <td>29066891</td>\n",
       "      <td>29066777</td>\n",
       "      <td>-</td>\n",
       "      <td>114</td>\n",
       "      <td>TATCAACATTGAAATGGCCGGAATCAGGAAAAATAGATACGCATGC...</td>\n",
       "      <td>...</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>4.913043</td>\n",
       "      <td>0.999999</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.234870</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>(29066777, 29066891)</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>chr21:CCT8:29069423:-|chr21:CCT8:29067705:-|lin</td>\n",
       "      <td>19</td>\n",
       "      <td>CCT8</td>\n",
       "      <td>CCT8</td>\n",
       "      <td>chr21</td>\n",
       "      <td>29069423</td>\n",
       "      <td>29067705</td>\n",
       "      <td>-</td>\n",
       "      <td>1718</td>\n",
       "      <td>AAGCCATTACAATCATTTTTGCAGCAGGATGCTGTACTTCTAGTTC...</td>\n",
       "      <td>...</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>5.052632</td>\n",
       "      <td>0.999988</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.212684</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>(29067705, 29069423)</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>chr21:CCT8:29070247:-|chr21:CCT8:29069502:-|lin</td>\n",
       "      <td>14</td>\n",
       "      <td>CCT8</td>\n",
       "      <td>CCT8</td>\n",
       "      <td>chr21</td>\n",
       "      <td>29070247</td>\n",
       "      <td>29069502</td>\n",
       "      <td>-</td>\n",
       "      <td>745</td>\n",
       "      <td>AAACAACTTCTCCAAGTGGTTGATAACCATTTTGTTCATTCCATTT...</td>\n",
       "      <td>...</td>\n",
       "      <td>2.642857</td>\n",
       "      <td>3.714286</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.168286</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>(29069502, 29070247)</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>chr21:CCT8:29073531:-|chr21:CCT8:29070337:-|lin</td>\n",
       "      <td>13</td>\n",
       "      <td>CCT8</td>\n",
       "      <td>CCT8</td>\n",
       "      <td>chr21</td>\n",
       "      <td>29073531</td>\n",
       "      <td>29070337</td>\n",
       "      <td>-</td>\n",
       "      <td>3194</td>\n",
       "      <td>TATGTTTCTATACACAGCCTCTTCTAATCCTGAAAAGTGTTTCGCT...</td>\n",
       "      <td>...</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>3.923077</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.202462</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>(29070337, 29073531)</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      refName_newR1  numReads geneR1A_uniq  \\\n",
       "50  chr21:CCT8:29060541:-|chr21:CCT8:29056552:-|lin        25         CCT8   \n",
       "51  chr21:CCT8:29061253:-|chr21:CCT8:29060660:-|lin        34         CCT8   \n",
       "52  chr21:CCT8:29061496:-|chr21:CCT8:29061417:-|lin        24         CCT8   \n",
       "53  chr21:CCT8:29062128:-|chr21:CCT8:29061567:-|lin        33         CCT8   \n",
       "54  chr21:CCT8:29063352:-|chr21:CCT8:29062556:-|lin        23         CCT8   \n",
       "55  chr21:CCT8:29066891:-|chr21:CCT8:29066777:-|lin        23         CCT8   \n",
       "56  chr21:CCT8:29069423:-|chr21:CCT8:29067705:-|lin        19         CCT8   \n",
       "57  chr21:CCT8:29070247:-|chr21:CCT8:29069502:-|lin        14         CCT8   \n",
       "58  chr21:CCT8:29073531:-|chr21:CCT8:29070337:-|lin        13         CCT8   \n",
       "\n",
       "   geneR1B_uniq chrR1A  juncPosR1A  juncPosR1B gene_strandR1A  intron_length  \\\n",
       "50         CCT8  chr21    29060541    29056552              -           3989   \n",
       "51         CCT8  chr21    29061253    29060660              -            593   \n",
       "52         CCT8  chr21    29061496    29061417              -             79   \n",
       "53         CCT8  chr21    29062128    29061567              -            561   \n",
       "54         CCT8  chr21    29063352    29062556              -            796   \n",
       "55         CCT8  chr21    29066891    29066777              -            114   \n",
       "56         CCT8  chr21    29069423    29067705              -           1718   \n",
       "57         CCT8  chr21    29070247    29069502              -            745   \n",
       "58         CCT8  chr21    29073531    29070337              -           3194   \n",
       "\n",
       "                                                seqR1  ...  ave_GC_run_R1  \\\n",
       "50  GGAGGCTTGGGCCCACCAGCTGGTTTTGCCATGATGATCTGATCCA...  ...       3.000000   \n",
       "51  CAGCTTCCAGCATGTCCTTTACAGCAGGGACTTCAGCCTCAATATC...  ...       2.823529   \n",
       "52  TCTTAATAGCATACTGTTCAAGTCCAGGACATGTCTCTCCATATGA...  ...       2.833333   \n",
       "53  TTTCTGTTGCTCCACCTCCGGGTACAAGACGTTTATCCCTTGTAAG...  ...       3.060606   \n",
       "54  AGTTTTACAAAGTCTTCGGAGATCCCATTTTGAGTTTAGCCTCACT...  ...       2.695652   \n",
       "55  TATCAACATTGAAATGGCCGGAATCAGGAAAAATAGATACGCATGC...  ...       2.000000   \n",
       "56  AAGCCATTACAATCATTTTTGCAGCAGGATGCTGTACTTCTAGTTC...  ...       2.000000   \n",
       "57  AAACAACTTCTCCAAGTGGTTGATAACCATTTTGTTCATTCCATTT...  ...       2.642857   \n",
       "58  TATGTTTCTATACACAGCCTCTTCTAATCCTGAAAAGTGTTTCGCT...  ...       3.000000   \n",
       "\n",
       "    ave_max_run_R1  junc_cdf_glmnet_constrained  emp.p_glmnet_constrained  \\\n",
       "50        4.000000                     0.999999                       0.0   \n",
       "51        4.647059                     1.000000                       0.0   \n",
       "52        3.250000                     1.000000                       0.0   \n",
       "53        3.060606                     1.000000                       0.0   \n",
       "54        3.782609                     0.999998                       0.0   \n",
       "55        4.913043                     0.999999                       0.0   \n",
       "56        5.052632                     0.999988                       0.0   \n",
       "57        3.714286                     1.000000                       0.0   \n",
       "58        3.923077                     1.000000                       0.0   \n",
       "\n",
       "    ave_entropyR1  exon_annR1A  exon_annR1B  both_ann             sort_junc  \\\n",
       "50       4.197200         True         True      True  (29056552, 29060541)   \n",
       "51       4.250088         True         True      True  (29060660, 29061253)   \n",
       "52       4.223083         True         True      True  (29061417, 29061496)   \n",
       "53       4.249121         True         True      True  (29061567, 29062128)   \n",
       "54       4.229609         True         True      True  (29062556, 29063352)   \n",
       "55       4.234870         True         True      True  (29066777, 29066891)   \n",
       "56       4.212684         True         True      True  (29067705, 29069423)   \n",
       "57       4.168286         True         True      True  (29069502, 29070247)   \n",
       "58       4.202462         True         True      True  (29070337, 29073531)   \n",
       "\n",
       "    splice_ann  \n",
       "50       False  \n",
       "51       False  \n",
       "52       False  \n",
       "53       False  \n",
       "54       False  \n",
       "55       False  \n",
       "56       False  \n",
       "57       False  \n",
       "58       False  \n",
       "\n",
       "[9 rows x 23 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df.geneR1A_uniq.values == \"CCT8\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e9b2d2ab-4d15-45f1-8a63-a07aba8b16c1",
   "metadata": {},
   "source": [
    "If you consider the [the Ensembl region in detail for CCT8](https://useast.ensembl.org/Homo_sapiens/Location/View?db=core;g=ENSG00000156261;r=21:29055802-29073805), especially the MANE Select Transcripts track, it matches quite well for many of the junctions expected. (I suggest looking under the `sort_junc` column almost all the way on the right in the table above.) These match up quite well with the following linear splice junctions from RJunBase database of human RNA splice junctions (ref. [Li. et al 2021](https://pubmed.ncbi.nlm.nih.gov/33179749/)):\n",
    "\n",
    "- http://www.rjunbase.org/rJunBase/search/detailInfoBefore?junctionId=CCT8_LS002&kind=ls  \n",
    "- http://www.rjunbase.org/rJunBase/search/detailInfoBefore?junctionId=CCT8_LS003&kind=ls  \n",
    "- http://www.rjunbase.org/rJunBase/search/detailInfoBefore?junctionId=CCT8_LS004&kind=ls  \n",
    "- http://www.rjunbase.org/rJunBase/search/detailInfoBefore?junctionId=CCT8_LS005&kind=ls  \n",
    "- http://www.rjunbase.org/rJunBase/search/detailInfoBefore?junctionId=CCT8_LS008&kind=ls  \n",
    "- http://www.rjunbase.org/rJunBase/search/detailInfoBefore?junctionId=CCT8_LS014&kind=ls  \n",
    "- http://www.rjunbase.org/rJunBase/search/detailInfoBefore?junctionId=CCT8_LS018&kind=ls  \n",
    "- http://www.rjunbase.org/rJunBase/search/detailInfoBefore?junctionId=CCT8_LS020&kind=ls  \n",
    "- http://www.rjunbase.org/rJunBase/search/detailInfoBefore?junctionId=CCT8_LS031&kind=ls  \n",
    "\n",
    "The `numReads` column, second from the left side of the table above, indicates many of these are well represented in the data and probably why deemed high-confidence. **Note all of them have both sides of the junction annotated**, indicated by the `True` for each in the `both_ann` column, that is to the left of the `sort_junc` column."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f138dac-973a-4cbc-980f-07a09f38e98b",
   "metadata": {},
   "source": [
    "-------\n",
    "\n",
    "Enjoy!"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
