{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "29f9c521-f794-4da6-a44f-ed03ee62be2f",
   "metadata": {},
   "source": [
    "# Run SICILIAN demonstration\n",
    "\n",
    "SICILIAN needs input data and the STAR aligner index for that reference genome. I have a test for STAR aligner with a single read that works on MyBinder, and so run that and then index will be here because that run makes the index as the third step."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc23e670-2be6-4a36-90c4-8bdfa4a0b68d",
   "metadata": {},
   "source": [
    "### Preparation STEP #1: Get input data and make STAR aligner index as part of aligning the data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7187ffa3-3481-4ae4-8ac1-1506e280f3e8",
   "metadata": {
    "tags": []
   },
   "source": [
    "Install the STAR Aligner to prepare and make a tiny dataset and then index that mock reference genome with STAR aligner and then map a read to it."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c0e7d808-1a4c-441a-b164-6fa16fc2c7b9",
   "metadata": {},
   "source": [
    "To do all that, we can run the Jupyter notebook file `STAR_aligner_run_on_MyBinder.ipynb` with the following command, I have added printing the steps to stderr (the pink highlighted lines) to better enable understanding the context of the steps without needing to actually open the notebook `STAR_aligner_run_on_MyBinder.ipynb`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "928fb930-4a8f-430f-9f3b-edbea2ad63f3",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Install the STAR Aligner to prepare\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Channels:\n",
      " - conda-forge\n",
      " - defaults\n",
      " - bioconda\n",
      "Platform: linux-64\n",
      "Collecting package metadata (repodata.json): done\n",
      "Solving environment: done\n",
      "\n",
      "\n",
      "==> WARNING: A newer version of conda exists. <==\n",
      "    current version: 24.3.0\n",
      "    latest version: 24.5.0\n",
      "\n",
      "Please update conda by running\n",
      "\n",
      "    $ conda update -n base -c conda-forge conda\n",
      "\n",
      "\n",
      "\n",
      "# All requested packages already installed.\n",
      "\n",
      "\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Run a test alignment to demonstrate it works\n",
      "1. Create a small FASTA file with a short sequence:\n",
      "2. Run `mkdir test_index` to make the directory to store the index."
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘test_index’: File exists\n",
      "\t/srv/conda/envs/notebook/bin/STAR-avx2 --runMode genomeGenerate --genomeFastaFiles test.fa --genomeDir test_index --genomeSAindexNbases 1\n",
      "\tSTAR version: 2.7.11b   compiled: 2024-03-19T08:38:59+0000 :/opt/conda/conda-bld/star_1710837244939/work/source\n",
      "Jun 06 20:16:20 ..... started STAR run\n",
      "Jun 06 20:16:20 ... starting to generate Genome files\n",
      "Jun 06 20:16:20 ... starting to sort Suffix Array. This may take a long time...\n",
      "Jun 06 20:16:20 ... sorting Suffix Array chunks and saving them to disk...\n",
      "Jun 06 20:16:20 ... loading chunks from disk, packing SA...\n",
      "Jun 06 20:16:20 ... finished generating suffix array\n",
      "Jun 06 20:16:20 ... generating Suffix Array index\n",
      "Jun 06 20:16:20 ... completed Suffix Array index\n",
      "Jun 06 20:16:20 ... writing Genome to disk ...\n",
      "Jun 06 20:16:20 ... writing Suffix Array to disk ...\n",
      "Jun 06 20:16:20 ... writing SAindex to disk\n",
      "Jun 06 20:16:20 ..... finished successfully\n",
      "\t/srv/conda/envs/notebook/bin/STAR-avx2 --runMode alignReads --genomeDir test_index --readFilesIn test.fq --outFileNamePrefix test_output\n",
      "\tSTAR version: 2.7.11b   compiled: 2024-03-19T08:38:59+0000 :/opt/conda/conda-bld/star_1710837244939/work/source\n",
      "Jun 06 20:16:20 ..... started STAR run\n",
      "Jun 06 20:16:20 ..... loading genome\n",
      "Jun 06 20:16:20 ..... started mapping\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "3. Generate a genome index using this small FASTA file:\n",
      "4. Create a small FASTQ file with a few reads from the test sequence:\n",
      "5. Run the alignment using the small test files:\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Jun 06 20:16:20 ..... finished mapping\n",
      "Jun 06 20:16:20 ..... finished successfully\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "SHOW THE FILES GENERATED BY ALL THAT AND THE ALIGNING STEP.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 412K\n",
      "drwxr-x--- 1 jovyan jovyan 4.0K Jun  6 20:16 \u001b[0m\u001b[01;34m.\u001b[0m/\n",
      "drwxr-xr-x 1 root   root   4.0K Jun  5 20:31 \u001b[01;34m..\u001b[0m/\n",
      "-rw-r--r-- 1 jovyan jovyan   25 Jun  6 20:15 a_tsv.tsv\n",
      "-rw-r--r-- 1 jovyan jovyan  220 Jan  6  2022 .bash_logout\n",
      "-rw-r--r-- 1 jovyan jovyan 3.7K Jan  6  2022 .bashrc\n",
      "drwxr-xr-x 3 jovyan jovyan 4.0K Jun  6 02:54 \u001b[01;34mbenchmarking\u001b[0m/\n",
      "drwxr-xr-x 1 jovyan jovyan 4.0K Jun  6 02:54 \u001b[01;34mbinder\u001b[0m/\n",
      "drwxr-xr-x 1 jovyan jovyan 4.0K Jun  6 18:10 \u001b[01;34m.cache\u001b[0m/\n",
      "drwxrwsr-x 1 jovyan jovyan 4.0K Jun  5 20:25 \u001b[01;34m.conda\u001b[0m/\n",
      "drwxr-xr-x 8 jovyan jovyan 4.0K Jun  6 02:54 \u001b[01;34m.git\u001b[0m/\n",
      "-rw-r--r-- 1 jovyan jovyan   94 Jun  6 02:54 .gitignore\n",
      "drwxr-xr-x 2 jovyan jovyan 4.0K Jun  6 20:15 \u001b[01;34m.ipynb_checkpoints\u001b[0m/\n",
      "drwxr-xr-x 1 jovyan jovyan 4.0K Jun  6 17:46 \u001b[01;34m.ipython\u001b[0m/\n",
      "drwxr-xr-x 3 jovyan jovyan 4.0K Jun  6 17:44 \u001b[01;34m.jupyter\u001b[0m/\n",
      "-rw-r--r-- 1 jovyan jovyan  20K Jun  6 20:16 .jupyter-server-log.txt\n",
      "-rw-r--r-- 1 jovyan jovyan  18K Jun  6 02:54 LICENSE\n",
      "drwxr-xr-x 1 jovyan jovyan 4.0K Jun  5 20:32 \u001b[01;34m.local\u001b[0m/\n",
      "-rw-r--r-- 1 jovyan jovyan  450 Jun  6 18:52 more_installs.R\n",
      "drwxr-xr-x 3 jovyan jovyan 4.0K Jun  6 17:44 \u001b[01;34m.npm\u001b[0m/\n",
      "-rw-r--r-- 1 jovyan jovyan  807 Jan  6  2022 .profile\n",
      "-rw-r--r-- 1 jovyan jovyan  26K Jun  6 02:54 README.md\n",
      "-rw-r--r-- 1 jovyan jovyan   98 Jun  6 02:54 requirements.txt\n",
      "-rw-r--r-- 1 jovyan jovyan 1.1K Jun  6 19:24 run_class_input.sh\n",
      "-rw-r--r-- 1 jovyan jovyan 1.2K Jun  6 19:24 run_GLM.sh\n",
      "-rw-r--r-- 1 jovyan jovyan 9.8K Jun  6 20:15 Run_SICILIAN_demo_on_MyBinder.ipynb\n",
      "-rw-r--r-- 1 jovyan jovyan 1.5K Jun  6 02:54 RUN_SICILIAN_post_process.sh\n",
      "drwxr-xr-x 1 jovyan jovyan 4.0K Jun  6 18:18 \u001b[01;34mscripts\u001b[0m/\n",
      "-rw-r--r-- 1 jovyan jovyan 141K Jun  6 02:54 SICILIAN.png\n",
      "-rw-r--r-- 1 jovyan jovyan 6.7K Jun  6 02:54 SICILIAN_post_process.py\n",
      "-rw-r--r-- 1 jovyan jovyan 9.6K Jun  6 18:13 SICILIAN.py\n",
      "-rw-r--r-- 1 jovyan jovyan  19K Jun  6 20:14 STAR_aligner_run_on_MyBinder.ipynb\n",
      "-rw-r--r-- 1 jovyan jovyan   72 Jun  6 20:16 test.fa\n",
      "-rw-r--r-- 1 jovyan jovyan   49 Jun  6 20:16 test.fq\n",
      "drwxr-xr-x 2 jovyan jovyan 4.0K Jun  6 20:16 \u001b[01;34mtest_index\u001b[0m/\n",
      "-rw-r--r-- 1 jovyan jovyan  481 Jun  6 20:16 test_outputAligned.out.sam\n",
      "-rw-r--r-- 1 jovyan jovyan 2.0K Jun  6 20:16 test_outputLog.final.out\n",
      "-rw-r--r-- 1 jovyan jovyan 3.7K Jun  6 20:16 test_outputLog.out\n",
      "-rw-r--r-- 1 jovyan jovyan  246 Jun  6 20:16 test_outputLog.progress.out\n",
      "-rw-r--r-- 1 jovyan jovyan    0 Jun  6 20:16 test_outputSJ.out.tab\n",
      "-rw-r--r-- 1 jovyan jovyan 1.2K Jun  6 20:16 Untitled.ipynb\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Examine results to see it worked\n",
      "Results in 'test_outputLog.final.out'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                 Started job on |\tJun 06 20:16:20\n",
      "                             Started mapping on |\tJun 06 20:16:20\n",
      "                                    Finished on |\tJun 06 20:16:20\n",
      "       Mapping speed, Million of reads per hour |\tinf\n",
      "\n",
      "                          Number of input reads |\t1\n",
      "                      Average input read length |\t19\n",
      "                                    UNIQUE READS:\n",
      "                   Uniquely mapped reads number |\t1\n",
      "                        Uniquely mapped reads % |\t100.00%\n",
      "                          Average mapped length |\t19.00\n",
      "                       Number of splices: Total |\t0\n",
      "            Number of splices: Annotated (sjdb) |\t0\n",
      "                       Number of splices: GT/AG |\t0\n",
      "                       Number of splices: GC/AG |\t0\n",
      "                       Number of splices: AT/AC |\t0\n",
      "               Number of splices: Non-canonical |\t0\n",
      "                      Mismatch rate per base, % |\t0.00%\n",
      "                         Deletion rate per base |\t0.00%\n",
      "                        Deletion average length |\t0.00\n",
      "                        Insertion rate per base |\t0.00%\n",
      "                       Insertion average length |\t0.00\n",
      "                             MULTI-MAPPING READS:\n",
      "        Number of reads mapped to multiple loci |\t0\n",
      "             % of reads mapped to multiple loci |\t0.00%\n",
      "        Number of reads mapped to too many loci |\t0\n",
      "             % of reads mapped to too many loci |\t0.00%\n",
      "                                  UNMAPPED READS:\n",
      "  Number of reads unmapped: too many mismatches |\t0\n",
      "       % of reads unmapped: too many mismatches |\t0.00%\n",
      "            Number of reads unmapped: too short |\t0\n",
      "                 % of reads unmapped: too short |\t0.00%\n",
      "                Number of reads unmapped: other |\t0\n",
      "                     % of reads unmapped: other |\t0.00%\n",
      "                                  CHIMERIC READS:\n",
      "                       Number of chimeric reads |\t0\n",
      "                            % of chimeric reads |\t0.00%\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Results in 'test_outputAligned.out.sam'"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "@HD\tVN:1.4\n",
      "@SQ\tSN:test_seq\tLN:61\n",
      "@PG\tID:STAR\tPN:STAR\tVN:2.7.11b\tCL:/srv/conda/envs/notebook/bin/STAR-avx2   --runMode alignReads      --genomeDir test_index   --readFilesIn test.fq      --outFileNamePrefix test_output\n",
      "@CO\tuser command line: /srv/conda/envs/notebook/bin/STAR-avx2 --runMode alignReads --genomeDir test_index --readFilesIn test.fq --outFileNamePrefix test_output\n",
      "read1\t0\ttest_seq\t19\t255\t19M\t*\t0\t0\tGTACCCAAGTACGAGTACG\tAAAAAAAAAAAAAAAAAAA\tNH:i:1\tHI:i:1\tAS:i:18\tnM:i:0\n"
     ]
    }
   ],
   "source": [
    "%run STAR_aligner_run_on_MyBinder.ipynb"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fac5bcee-fbe3-4596-83a0-cead799ba1b2",
   "metadata": {},
   "source": [
    "### Preparation STEP #2: Use SICILIAN's `create_annotator.py` script to prepare for SICILIAN's main script to run\n",
    "\n",
    "The annotator pickle file for a GTF file needs to be made only once by the `create_annotator.py` script in the SICILIAN scripts directory."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43c4ec57-a418-4d20-8b1e-71df9da61aba",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "s='''test_seq\\tSTAR\\tgene\\t1\\t60\\t.\\t+\\t.\\tgene_id \"TEST_GENE_1\"; gene_name \"test_gene_1\";\n",
    "test_seq\\tSTAR\\ttranscript\\t1\\t20\\t.\\t+\\t.\\tgene_id \"TEST_GENE_1\"; transcript_id \"TRANSCRIPT_1\"; gene_name \"test_gene_1\";\n",
    "test_seq\\tSTAR\\texon\\t1\\t20\\t.\\t+\\t.\\tgene_id \"TEST_GENE_1\"; transcript_id \"TRANSCRIPT_1\"; exon_number \"1\"; gene_name \"test_gene_1\";\n",
    "test_seq\\tSTAR\\ttranscript\\t21\\t40\\t.\\t+\\t.\\tgene_id \"TEST_GENE_1\"; transcript_id \"TRANSCRIPT_2\"; gene_name \"test_gene_1\";\n",
    "test_seq\\tSTAR\\texon\\t21\\t40\\t.\\t+\\t.\\tgene_id \"TEST_GENE_1\"; transcript_id \"TRANSCRIPT_2\"; exon_number \"1\"; gene_name \"test_gene_1\";\n",
    "test_seq\\tSTAR\\ttranscript\\t41\\t60\\t.\\t+\\t.\\tgene_id \"TEST_GENE_1\"; transcript_id \"TRANSCRIPT_3\"; gene_name \"test_gene_1\";\n",
    "test_seq\\tSTAR\\texon\\t41\\t60\\t.\\t+\\t.\\tgene_id \"TEST_GENE_1\"; transcript_id \"TRANSCRIPT_3\"; exon_number \"1\"; gene_name \"test_gene_1\";\n",
    "test_seq\\tSTAR\\tgene\\t1\\t60\\t.\\t+\\t.\\tgene_id \"TEST_GENE_2\"; gene_name \"test_gene_2\";\n",
    "test_seq\\tSTAR\\ttranscript\\t1\\t30\\t.\\t+\\t.\\tgene_id \"TEST_GENE_2\"; transcript_id \"TRANSCRIPT_4\"; gene_name \"test_gene_2\";\n",
    "test_seq\\tSTAR\\texon\\t1\\t10\\t.\\t+\\t.\\tgene_id \"TEST_GENE_2\"; transcript_id \"TRANSCRIPT_4\"; exon_number \"1\"; gene_name \"test_gene_2\";\n",
    "test_seq\\tSTAR\\texon\\t21\\t30\\t.\\t+\\t.\\tgene_id \"TEST_GENE_2\"; transcript_id \"TRANSCRIPT_4\"; exon_number \"2\"; gene_name \"test_gene_2\";\n",
    "test_seq\\tSTAR\\ttranscript\\t31\\t60\\t.\\t+\\t.\\tgene_id \"TEST_GENE_2\"; transcript_id \"TRANSCRIPT_5\"; gene_name \"test_gene_2\";\n",
    "test_seq\\tSTAR\\texon\\t31\\t40\\t.\\t+\\t.\\tgene_id \"TEST_GENE_2\"; transcript_id \"TRANSCRIPT_5\"; exon_number \"1\"; gene_name \"test_gene_2\";\n",
    "test_seq\\tSTAR\\texon\\t51\\t60\\t.\\t+\\t.\\tgene_id \"TEST_GENE_2\"; transcript_id \"TRANSCRIPT_5\"; exon_number \"2\"; gene_name \"test_gene_2\";'''\n",
    "%store s >test_gtf_file.gtf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ac082e62-c4c8-4435-b861-c6f243797122",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%run scripts/create_annotator.py -g test_gtf_file.gtf -a testsq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a5de9db-7a19-49b9-bfd3-c5a7ec0787bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Preparation STEP #3: set the settings in SICILIAN to reflect all this set up."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae7adba2-3983-4999-9c5f-143a296b48e5",
   "metadata": {},
   "source": [
    "-----\n",
    "\n",
    "## Run SICILIAN\n",
    "\n",
    "Run the main SICILIAN script by running the following cell. (If you were on the command line the following command would be more something along the lines of `python SICILIAN.py`. Here I'm using Jupyter's convenient magic that is more feature-rich.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9745a6ec-07fd-4593-9691-0dff2cc4011d",
   "metadata": {},
   "outputs": [],
   "source": [
    "%run SICILIAN.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f2be2458-ea38-4fc3-a422-56026740dc94",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
